--- ../src-base/minecraft/net/minecraft/entity/projectile/EntityFireball.java
+++ ../src-work/minecraft/net/minecraft/entity/projectile/EntityFireball.java
@@ -15,6 +15,8 @@
 import net.minecraft.util.Vec3;
 import net.minecraft.world.World;
 
+import org.bukkit.event.entity.EntityDamageByEntityEvent; // CraftBukkit
+
 public abstract class EntityFireball extends Entity
 {
     private int field_145795_e = -1;
@@ -28,6 +30,8 @@
     public double accelerationX;
     public double accelerationY;
     public double accelerationZ;
+    public float bukkitYield = 1; // CraftBukkit
+    public boolean isIncendiary = true; // CraftBukkit
     private static final String __OBFID = "CL_00001717";
 
     public EntityFireball(World par1World)
@@ -63,20 +67,28 @@
     {
         super(par1World);
         this.shootingEntity = par2EntityLivingBase;
+        this.projectileSource = (org.bukkit.entity.LivingEntity) par2EntityLivingBase.getBukkitEntity(); // CraftBukkit
         this.setSize(1.0F, 1.0F);
         this.setLocationAndAngles(par2EntityLivingBase.posX, par2EntityLivingBase.posY, par2EntityLivingBase.posZ, par2EntityLivingBase.rotationYaw, par2EntityLivingBase.rotationPitch);
         this.setPosition(this.posX, this.posY, this.posZ);
         this.yOffset = 0.0F;
         this.motionX = this.motionY = this.motionZ = 0.0D;
-        par3 += this.rand.nextGaussian() * 0.4D;
-        par5 += this.rand.nextGaussian() * 0.4D;
-        par7 += this.rand.nextGaussian() * 0.4D;
-        double d3 = (double)MathHelper.sqrt_double(par3 * par3 + par5 * par5 + par7 * par7);
-        this.accelerationX = par3 / d3 * 0.1D;
-        this.accelerationY = par5 / d3 * 0.1D;
-        this.accelerationZ = par7 / d3 * 0.1D;
+        // CraftBukkit start - Added setDirection method
+        this.setDirection(par3, par5, par7);
     }
 
+    public void setDirection(double d0, double d1, double d2)
+    {
+        // CraftBukkit end
+        d0 += this.rand.nextGaussian() * 0.4D;
+        d1 += this.rand.nextGaussian() * 0.4D;
+        d2 += this.rand.nextGaussian() * 0.4D;
+        double d3 = (double) MathHelper.sqrt_double(d0 * d0 + d1 * d1 + d2 * d2);
+        this.accelerationX = d0 / d3 * 0.1D;
+        this.accelerationY = d1 / d3 * 0.1D;
+        this.accelerationZ = d2 / d3 * 0.1D;
+    }
+
     // JAVADOC METHOD $$ func_70071_h_
     public void onUpdate()
     {
@@ -142,7 +154,7 @@
 
                     if (movingobjectposition1 != null)
                     {
-                        double d1 = vec3.distanceTo(movingobjectposition1.hitVec);
+                        double d1 = vec3.squareDistanceTo(movingobjectposition1.hitVec); // CraftBukkit - distance efficiency
 
                         if (d1 < d0 || d0 == 0.0D)
                         {
@@ -161,6 +173,14 @@
             if (movingobjectposition != null)
             {
                 this.onImpact(movingobjectposition);
+
+                // CraftBukkit start
+                if (this.isDead)
+                {
+                    org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileHitEvent(this);
+                }
+
+                // CraftBukkit end
             }
 
             this.posX += this.motionX;
@@ -232,7 +252,10 @@
         par1NBTTagCompound.setShort("zTile", (short)this.field_145794_g);
         par1NBTTagCompound.setByte("inTile", (byte)Block.getIdFromBlock(this.field_145796_h));
         par1NBTTagCompound.setByte("inGround", (byte)(this.inGround ? 1 : 0));
-        par1NBTTagCompound.setTag("direction", this.newDoubleNBTList(new double[] {this.motionX, this.motionY, this.motionZ}));
+        // CraftBukkit - Fix direction being mismapped to invalid variables
+        par1NBTTagCompound.setTag("power", this.newDoubleNBTList(new double[] { this.accelerationX, this.accelerationY, this.accelerationZ}));
+        // Spigot - Support vanilla's direction tag
+        par1NBTTagCompound.setTag("direction", this.newDoubleNBTList(new double[] { this.motionX, this.motionY, this.motionZ}));
     }
 
     // JAVADOC METHOD $$ func_70037_a
@@ -244,8 +267,17 @@
         this.field_145796_h = Block.getBlockById(par1NBTTagCompound.getByte("inTile") & 255);
         this.inGround = par1NBTTagCompound.getByte("inGround") == 1;
 
-        if (par1NBTTagCompound.hasKey("direction", 9))
+        // CraftBukkit start - direction -> power
+        if (par1NBTTagCompound.hasKey("power", 9))
         {
+            NBTTagList nbttaglist = par1NBTTagCompound.getTagList("power", 6);
+            this.accelerationX = nbttaglist.func_150309_d(0);
+            this.accelerationY = nbttaglist.func_150309_d(1);
+            this.accelerationZ = nbttaglist.func_150309_d(2);
+            // CraftBukkit end
+        }
+        else if (par1NBTTagCompound.hasKey("direction", 9))     // Spigot - Support vanilla's direction tag
+        {
             NBTTagList nbttaglist = par1NBTTagCompound.getTagList("direction", 6);
             this.motionX = nbttaglist.func_150309_d(0);
             this.motionY = nbttaglist.func_150309_d(1);
@@ -281,6 +313,16 @@
 
             if (par1DamageSource.getEntity() != null)
             {
+                // CraftBukkit start
+                EntityDamageByEntityEvent event = new EntityDamageByEntityEvent(par1DamageSource.getEntity().getBukkitEntity(), this.getBukkitEntity(), org.bukkit.event.entity.EntityDamageEvent.DamageCause.ENTITY_ATTACK, par2);
+                worldObj.getServer().getPluginManager().callEvent(event);
+
+                if (event.isCancelled())
+                {
+                    return false;
+                }
+
+                // CraftBukkit end
                 Vec3 vec3 = par1DamageSource.getEntity().getLookVec();
 
                 if (vec3 != null)
