--- ../src-base/minecraft/cpw/mods/fml/common/network/handshake/NetworkDispatcher.java
+++ ../src-work/minecraft/cpw/mods/fml/common/network/handshake/NetworkDispatcher.java
@@ -201,7 +201,7 @@
         }
         else
         {
-            FMLLog.info("Unexpected packet during modded negotiation - assuming vanilla or keepalives : %s", msg.getClass().getName());
+            //FMLLog.info("Unexpected packet during modded negotiation - assuming vanilla or keepalives : %s", msg.getClass().getName()); // MCPC+ - unneeded spam
         }
         return false;
     }
@@ -453,7 +453,7 @@
         // Stop the epic channel closed spam at close
         if (!(cause instanceof ClosedChannelException))
         {
-            FMLLog.log(Level.ERROR, cause, "NetworkDispatcher exception");
+            //FMLLog.log(Level.ERROR, cause, "NetworkDispatcher exception"); // MCPC+ - disable unneeded spam
         }
         super.exceptionCaught(ctx, cause);
     }
