--- ../src-base/minecraft/net/minecraft/world/biome/BiomeGenBase.java
+++ ../src-work/minecraft/net/minecraft/world/biome/BiomeGenBase.java
@@ -62,7 +62,7 @@
     protected static final BiomeGenBase.Height field_150598_l = new BiomeGenBase.Height(0.2F, 0.3F);
     protected static final BiomeGenBase.Height field_150599_m = new BiomeGenBase.Height(-0.2F, 0.1F);
     // JAVADOC FIELD $$ field_76773_a
-    private static final BiomeGenBase[] biomeList = new BiomeGenBase[256];
+    public static final BiomeGenBase[] biomeList = new BiomeGenBase[256]; // MCPC+ - private -> public
     public static final Set field_150597_n = Sets.newHashSet();
     public static final BiomeGenBase ocean = (new BiomeGenOcean(0)).setColor(112).setBiomeName("Ocean").func_150570_a(field_150595_c);
     public static final BiomeGenBase plains = (new BiomeGenPlains(1)).setColor(9286496).setBiomeName("Plains");
