--- ../src-base/minecraft/net/minecraft/entity/passive/EntitySheep.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntitySheep.java
@@ -32,6 +32,12 @@
 import net.minecraft.world.World;
 import net.minecraftforge.common.IShearable;
 
+// CraftBukkit start
+import net.minecraft.inventory.InventoryCraftResult;
+import org.bukkit.event.entity.SheepRegrowWoolEvent;
+import org.bukkit.event.player.PlayerShearEntityEvent;
+// CraftBukkit end
+
 public class EntitySheep extends EntityAnimal implements IShearable
 {
     private final InventoryCrafting field_90016_e = new InventoryCrafting(new Container()
@@ -63,6 +69,7 @@
         this.tasks.addTask(8, new EntityAILookIdle(this));
         this.field_90016_e.setInventorySlotContents(0, new ItemStack(Items.dye, 1, 0));
         this.field_90016_e.setInventorySlotContents(1, new ItemStack(Items.dye, 1, 0));
+        this.field_90016_e.resultInventory = new InventoryCraftResult(); // CraftBukkit - add result slot for event
     }
 
     protected boolean isAIEnabled()
@@ -101,10 +108,16 @@
 
     protected void dropFewItems(boolean par1, int par2)
     {
+        // CraftBukkit start - Whole method
+        java.util.List<org.bukkit.inventory.ItemStack> loot = new java.util.ArrayList<org.bukkit.inventory.ItemStack>();
+
         if (!this.getSheared())
         {
-            this.entityDropItem(new ItemStack(Item.getItemFromBlock(Blocks.wool), 1, this.getFleeceColor()), 0.0F);
+            loot.add(new org.bukkit.inventory.ItemStack(org.bukkit.Material.WOOL, 1, (short) 0, (byte) this.getFleeceColor()));
         }
+
+        org.bukkit.craftbukkit.event.CraftEventFactory.callEntityDeathEvent(this, loot);
+        // CraftBukkit end
     }
 
     protected Item getDropItem()
@@ -127,6 +140,38 @@
 
     public boolean interact(EntityPlayer par1EntityPlayer)
     {
+        ItemStack itemstack = par1EntityPlayer.inventory.getCurrentItem();
+
+        if (itemstack != null && itemstack.getItem() == Items.shears && !this.getSheared() && !this.isChild())
+        {
+            if (!this.worldObj.isRemote)
+            {
+                // CraftBukkit start
+                PlayerShearEntityEvent event = new PlayerShearEntityEvent((org.bukkit.entity.Player) par1EntityPlayer.getBukkitEntity(), this.getBukkitEntity());
+                this.worldObj.getServer().getPluginManager().callEvent(event);
+
+                if (event.isCancelled())
+                {
+                    return false;
+                }
+
+                // CraftBukkit end
+                this.setSheared(true);
+                int i = 1 + this.rand.nextInt(3);
+
+                for (int j = 0; j < i; ++j)
+                {
+                    EntityItem entityitem = this.entityDropItem(new ItemStack(Item.getItemFromBlock(Blocks.wool), 1, this.getFleeceColor()), 1.0F);
+                    entityitem.motionY += (double)(this.rand.nextFloat() * 0.05F);
+                    entityitem.motionX += (double)((this.rand.nextFloat() - this.rand.nextFloat()) * 0.1F);
+                    entityitem.motionZ += (double)((this.rand.nextFloat() - this.rand.nextFloat()) * 0.1F);
+                }
+            }
+
+            itemstack.damageItem(1, par1EntityPlayer);
+            this.playSound("mob.sheep.shear", 1.0F, 1.0F);
+        }
+
         return super.interact(par1EntityPlayer);
     }
 
@@ -231,8 +276,17 @@
 
     public void eatGrassBonus()
     {
-        this.setSheared(false);
+        // CraftBukkit start
+        SheepRegrowWoolEvent event = new SheepRegrowWoolEvent((org.bukkit.entity.Sheep) this.getBukkitEntity());
+        this.worldObj.getServer().getPluginManager().callEvent(event);
 
+        if (!event.isCancelled())
+        {
+            this.setSheared(false);
+        }
+
+        // CraftBukkit end
+
         if (this.isChild())
         {
             this.addGrowth(60);
