--- ../src-base/minecraft/net/minecraft/entity/passive/EntityChicken.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntityChicken.java
@@ -18,6 +18,8 @@
 import net.minecraft.item.ItemStack;
 import net.minecraft.world.World;
 
+import org.bukkit.craftbukkit.util.CraftMagicNumbers; // CraftBukkit
+
 public class EntityChicken extends EntityAnimal
 {
     public float field_70886_e;
@@ -131,21 +133,26 @@
     // JAVADOC METHOD $$ func_70628_a
     protected void dropFewItems(boolean par1, int par2)
     {
+        // CraftBukkit start - Whole method
+        java.util.List<org.bukkit.inventory.ItemStack> loot = new java.util.ArrayList<org.bukkit.inventory.ItemStack>();
         int j = this.rand.nextInt(3) + this.rand.nextInt(1 + par2);
 
-        for (int k = 0; k < j; ++k)
+        if (j > 0)
         {
-            this.func_145779_a(Items.feather, 1);
+            loot.add(new org.bukkit.inventory.ItemStack(CraftMagicNumbers.getMaterial(Items.feather), j));
         }
 
         if (this.isBurning())
         {
-            this.func_145779_a(Items.cooked_chicken, 1);
+            loot.add(new org.bukkit.inventory.ItemStack(CraftMagicNumbers.getMaterial(Items.cooked_chicken), 1));
         }
         else
         {
-            this.func_145779_a(Items.chicken, 1);
+            loot.add(new org.bukkit.inventory.ItemStack(CraftMagicNumbers.getMaterial(Items.chicken), 1));
         }
+
+        org.bukkit.craftbukkit.event.CraftEventFactory.callEntityDeathEvent(this, loot);
+        // CraftBukkit end
     }
 
     public EntityChicken createChild(EntityAgeable par1EntityAgeable)
