--- ../src-base/minecraft/org/bukkit/World.java
+++ ../src-work/minecraft/org/bukkit/World.java
@@ -1249,6 +1249,12 @@
             return lookup.get(id);
         }
 
+        // MCPC+ start - allow forge to register environments
+        public static void registerEnvironment(Environment env) {
+            lookup.put(env.getId(),env);
+        }
+        // MCPC+ end
+
         static {
             for (Environment env : values()) {
                 lookup.put(env.getId(), env);
