--- ../src-base/minecraft/net/minecraft/tileentity/TileEntity.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntity.java
@@ -22,15 +22,18 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import org.spigotmc.CustomTimingsHandler; // Spigot
+import org.bukkit.inventory.InventoryHolder; // CraftBukkit
+
 public class TileEntity
 {
     private static final Logger logger = LogManager.getLogger();
     // JAVADOC FIELD $$ field_145855_i
     private static Map nameToClassMap = new HashMap();
     // JAVADOC FIELD $$ field_145853_j
-    private static Map classToNameMap = new HashMap();
+    public static Map classToNameMap = new HashMap(); // MCPC+ - private -> public
     // JAVADOC FIELD $$ field_145850_b
-    protected World worldObj;
+    public World worldObj; // CraftBukkit - protected -> public
     public int xCoord;
     public int yCoord;
     public int zCoord;
@@ -38,6 +41,7 @@
     public int blockMetadata = -1;
     // JAVADOC FIELD $$ field_145854_h
     public Block blockType;
+    public CustomTimingsHandler tickTimer = org.bukkit.craftbukkit.SpigotTimings.getTileEntityTimings(this); // Spigot
     private static final String __OBFID = "CL_00000340";
 
     // JAVADOC METHOD $$ func_145826_a
@@ -300,6 +304,20 @@
         addMapping(TileEntityFlowerPot.class, "FlowerPot");
     }
 
+    // CraftBukkit start
+    public InventoryHolder getOwner()
+    {
+        org.bukkit.block.BlockState state = worldObj.getWorld().getBlockAt(xCoord, yCoord, zCoord).getState();
+
+        if (state instanceof InventoryHolder)
+        {
+            return (InventoryHolder) state;
+        }
+
+        return null;
+    }
+    // CraftBukkit end
+
     // -- BEGIN FORGE PATCHES --
     /**
      * Determines if this TileEntity requires update calls.
