--- ../src-base/minecraft/net/minecraft/tileentity/TileEntityChest.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntityChest.java
@@ -15,6 +15,12 @@
 import net.minecraft.nbt.NBTTagList;
 import net.minecraft.util.AxisAlignedBB;
 
+// CraftBukkit start
+import net.minecraft.init.Blocks;
+import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
+// CraftBukkit end
+
 public class TileEntityChest extends TileEntity implements IInventory
 {
     private ItemStack[] field_145985_p = new ItemStack[36];
@@ -31,6 +37,36 @@
     private String field_145981_s;
     private static final String __OBFID = "CL_00000346";
 
+    // CraftBukkit start
+    public List<HumanEntity> transaction = new java.util.ArrayList<HumanEntity>();
+    private int maxStack = MAX_STACK;
+
+    public ItemStack[] getContents()
+    {
+        return this.field_145985_p;
+    }
+
+    public void onOpen(CraftHumanEntity who)
+    {
+        transaction.add(who);
+    }
+
+    public void onClose(CraftHumanEntity who)
+    {
+        transaction.remove(who);
+    }
+
+    public List<HumanEntity> getViewers()
+    {
+        return transaction;
+    }
+
+    public void setMaxStackSize(int size)
+    {
+        maxStack = size;
+    }
+    // CraftBukkit end
+
     public TileEntityChest()
     {
         this.field_145982_r = -1;
@@ -180,12 +216,17 @@
     // JAVADOC METHOD $$ func_70297_j_
     public int getInventoryStackLimit()
     {
-        return 64;
+        return maxStack; // CraftBukkit
     }
 
     // JAVADOC METHOD $$ func_70300_a
     public boolean isUseableByPlayer(EntityPlayer par1EntityPlayer)
     {
+        if (this.field_145850_b == null)
+        {
+            return true;    // CraftBukkit
+        }
+
         return this.field_145850_b.func_147438_o(this.field_145851_c, this.field_145848_d, this.field_145849_e) != this ? false : par1EntityPlayer.getDistanceSq((double)this.field_145851_c + 0.5D, (double)this.field_145848_d + 0.5D, (double)this.field_145849_e + 0.5D) <= 64.0D;
     }
 
@@ -296,6 +337,12 @@
     public void func_145845_h()
     {
         super.func_145845_h();
+
+        if (this.field_145850_b == null)
+        {
+            return;    // CraftBukkit
+        }
+
         this.func_145979_i();
         ++this.field_145983_q;
         float f;
@@ -410,8 +457,28 @@
             this.field_145987_o = 0;
         }
 
+        int oldPower = Math.max(0, Math.min(15, this.field_145987_o)); // CraftBukkit - Get power before new viewer is added
         ++this.field_145987_o;
+
+        if (this.field_145850_b == null)
+        {
+            return;    // CraftBukkit
+        }
+
         this.field_145850_b.func_147452_c(this.field_145851_c, this.field_145848_d, this.field_145849_e, this.func_145838_q(), 1, this.field_145987_o);
+
+        // CraftBukkit start - Call redstone event
+        if (this.func_145838_q() == Blocks.trapped_chest)
+        {
+            int newPower = Math.max(0, Math.min(15, this.field_145987_o));
+
+            if (oldPower != newPower)
+            {
+                org.bukkit.craftbukkit.event.CraftEventFactory.callRedstoneChange(field_145850_b, this.field_145851_c, this.field_145848_d, this.field_145849_e, oldPower, newPower);
+            }
+        }
+
+        // CraftBukkit end
         this.field_145850_b.func_147459_d(this.field_145851_c, this.field_145848_d, this.field_145849_e, this.func_145838_q());
         this.field_145850_b.func_147459_d(this.field_145851_c, this.field_145848_d - 1, this.field_145849_e, this.func_145838_q());
     }
@@ -420,8 +487,28 @@
     {
         if (this.func_145838_q() instanceof BlockChest)
         {
+            int oldPower = Math.max(0, Math.min(15, this.field_145987_o)); // CraftBukkit - Get power before new viewer is added
             --this.field_145987_o;
+
+            if (this.field_145850_b == null)
+            {
+                return;    // CraftBukkit
+            }
+
             this.field_145850_b.func_147452_c(this.field_145851_c, this.field_145848_d, this.field_145849_e, this.func_145838_q(), 1, this.field_145987_o);
+
+            // CraftBukkit start - Call redstone event
+            if (this.func_145838_q() == Blocks.trapped_chest)
+            {
+                int newPower = Math.max(0, Math.min(15, this.field_145987_o));
+
+                if (oldPower != newPower)
+                {
+                    org.bukkit.craftbukkit.event.CraftEventFactory.callRedstoneChange(field_145850_b, this.field_145851_c, this.field_145848_d, this.field_145849_e, oldPower, newPower);
+                }
+            }
+
+            // CraftBukkit end
             this.field_145850_b.func_147459_d(this.field_145851_c, this.field_145848_d, this.field_145849_e, this.func_145838_q());
             this.field_145850_b.func_147459_d(this.field_145851_c, this.field_145848_d - 1, this.field_145849_e, this.func_145838_q());
         }
