--- ../src-base/minecraft/net/minecraft/tileentity/TileEntity.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntity.java
@@ -22,18 +22,22 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import org.spigotmc.CustomTimingsHandler; // Spigot
+import org.bukkit.inventory.InventoryHolder; // CraftBukkit
+
 public class TileEntity
 {
     private static final Logger field_145852_a = LogManager.getLogger();
     private static Map field_145855_i = new HashMap();
-    private static Map field_145853_j = new HashMap();
-    protected World field_145850_b;
+    public static Map field_145853_j = new HashMap(); // MCPC+ - private -> public
+    public World field_145850_b; // CraftBukkit - protected -> public
     public int field_145851_c;
     public int field_145848_d;
     public int field_145849_e;
     protected boolean field_145846_f;
     public int field_145847_g = -1;
     public Block field_145854_h;
+    public CustomTimingsHandler tickTimer = org.bukkit.craftbukkit.SpigotTimings.getTileEntityTimings(this); // Spigot
     private static final String __OBFID = "CL_00000340";
 
     public static void func_145826_a(Class p_145826_0_, String p_145826_1_)
@@ -283,6 +287,20 @@
         func_145826_a(TileEntityFlowerPot.class, "FlowerPot");
     }
 
+    // CraftBukkit start
+    public InventoryHolder getOwner()
+    {
+        org.bukkit.block.BlockState state = field_145850_b.getWorld().getBlockAt(field_145851_c, field_145848_d, field_145849_e).getState();
+
+        if (state instanceof InventoryHolder)
+        {
+            return (InventoryHolder) state;
+        }
+
+        return null;
+    }
+    // CraftBukkit end
+
     // -- BEGIN FORGE PATCHES --
     /**
      * Determines if this TileEntity requires update calls.
