--- ../src-base/minecraft/net/minecraft/tileentity/TileEntitySkull.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntitySkull.java
@@ -57,6 +57,13 @@
         this.field_145910_i = p_145903_1_;
     }
 
+    // CraftBukkit start
+    public int getRotation()
+    {
+        return this.field_145910_i;
+    }
+    // CraftBukkit end
+
     @SideOnly(Side.CLIENT)
     public int func_145906_b()
     {
@@ -67,4 +74,12 @@
     {
         return this.field_145909_j;
     }
+
+    // MCPC+ start
+    @Override
+    public boolean canUpdate()
+    {
+        return false;
+    }
+    // MCPC+ end
 }
