--- ../src-base/minecraft/net/minecraft/entity/EntityLiving.java
+++ ../src-work/minecraft/net/minecraft/entity/EntityLiving.java
@@ -38,6 +38,13 @@
 import cpw.mods.fml.common.eventhandler.Event.Result;
 import net.minecraftforge.event.ForgeEventFactory;
 
+// CraftBukkit start
+import net.minecraft.entity.player.EntityPlayerMP;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityUnleashEvent;
+import org.bukkit.event.entity.EntityUnleashEvent.UnleashReason;
+// CraftBukkit end
+
 public abstract class EntityLiving extends EntityLivingBase
 {
     // JAVADOC FIELD $$ field_70757_a
@@ -50,19 +57,19 @@
     private EntityJumpHelper jumpHelper;
     private EntityBodyHelper bodyHelper;
     private PathNavigate navigator;
-    public final EntityAITasks tasks;
-    public final EntityAITasks targetTasks;
+    protected final EntityAITasks tasks;
+    protected final EntityAITasks targetTasks;
     // JAVADOC FIELD $$ field_70696_bz
     private EntityLivingBase attackTarget;
     private EntitySenses senses;
     // JAVADOC FIELD $$ field_82182_bS
     private ItemStack[] equipment = new ItemStack[5];
     // JAVADOC FIELD $$ field_82174_bp
-    protected float[] equipmentDropChances = new float[5];
+    public float[] equipmentDropChances = new float[5]; // CraftBukkit - protected -> public
     // JAVADOC FIELD $$ field_82172_bs
-    private boolean canPickUpLoot;
+    public boolean canPickUpLoot; // CraftBukkit - private -> public
     // JAVADOC FIELD $$ field_82179_bU
-    private boolean persistenceRequired;
+    public boolean persistenceRequired; // CraftBukkit - private -> public
     protected float defaultPitch;
     // JAVADOC FIELD $$ field_70776_bF
     private Entity currentTarget;
@@ -336,9 +343,23 @@
     public void readEntityFromNBT(NBTTagCompound par1NBTTagCompound)
     {
         super.readEntityFromNBT(par1NBTTagCompound);
-        this.setCanPickUpLoot(par1NBTTagCompound.getBoolean("CanPickUpLoot"));
-        this.persistenceRequired = par1NBTTagCompound.getBoolean("PersistenceRequired");
+        // CraftBukkit start - If looting or persistence is false only use it if it was set after we started using it
+        boolean data = par1NBTTagCompound.getBoolean("CanPickUpLoot");
 
+        if (isLevelAtLeast(par1NBTTagCompound, 1) || data)
+        {
+            this.canPickUpLoot = data;
+        }
+
+        data = par1NBTTagCompound.getBoolean("PersistenceRequired");
+
+        if (isLevelAtLeast(par1NBTTagCompound, 1) || data)
+        {
+            this.persistenceRequired = data;
+        }
+
+        // CraftBukkit end
+
         if (par1NBTTagCompound.func_150297_b("CustomName", 8) && par1NBTTagCompound.getString("CustomName").length() > 0)
         {
             this.setCustomNameTag(par1NBTTagCompound.getString("CustomName"));
@@ -551,15 +572,41 @@
                     this.entityAge = 0;
                 }
             }
+            else // MCPC+ start - Despawn entities when a player isn't in the world
+            {
+                if (this.worldObj.mcpcConfig.getBoolean("despawn-immediate", true) && this.canDespawn()) // Use flag for the assumption that being in another dimension entirely means no player is anywhere near
+                {
+                    this.despawn("No Player : Immediate");
+                }
+                else if (this.entityAge > 600 && this.rand.nextInt(800) == 0 && this.canDespawn()) // Otherwise, use the same logic as above
+                {
+                    this.despawn("No Player : Random Aged");
+                }
+            } // MCPC+ end
         }
     }
 
+    // MCPC+ start
+    private void despawn(String reason) {
+        this.setDead();
+        za.co.mcportcentral.MCPCHooks.logEntityDespawn(this, reason);
+    }
+    // MCPC+ end
+
     protected void updateAITasks()
     {
         ++this.entityAge;
         this.worldObj.theProfiler.startSection("checkDespawn");
         this.despawnEntity();
         this.worldObj.theProfiler.endSection();
+
+        // Spigot Start
+        if (this.fromMobSpawner)
+        {
+            return;
+        }
+
+        // Spigot End
         this.worldObj.theProfiler.startSection("sensing");
         this.senses.clearSensingCache();
         this.worldObj.theProfiler.endSection();
@@ -1052,6 +1099,14 @@
     {
         if (this.getLeashed() && this.getLeashedToEntity() == par1EntityPlayer)
         {
+            // CraftBukkit start
+            if (CraftEventFactory.callPlayerUnleashEntityEvent(this, par1EntityPlayer).isCancelled())
+            {
+                ((EntityPlayerMP) par1EntityPlayer).playerNetServerHandler.func_147359_a(new S1BPacketEntityAttach(1, this, this.getLeashedToEntity()));
+                return false;
+            }
+
+            // CraftBukkit end
             this.clearLeashed(true, !par1EntityPlayer.capabilities.isCreativeMode);
             return true;
         }
@@ -1063,6 +1118,14 @@
             {
                 if (!(this instanceof EntityTameable) || !((EntityTameable)this).isTamed())
                 {
+                    // CraftBukkit start
+                    if (CraftEventFactory.callPlayerLeashEntityEvent(this, par1EntityPlayer, par1EntityPlayer).isCancelled())
+                    {
+                        ((EntityPlayerMP) par1EntityPlayer).playerNetServerHandler.func_147359_a(new S1BPacketEntityAttach(1, this, this.getLeashedToEntity()));
+                        return false;
+                    }
+
+                    // CraftBukkit end
                     this.setLeashedToEntity(par1EntityPlayer, true);
                     --itemstack.stackSize;
                     return true;
@@ -1070,6 +1133,14 @@
 
                 if (par1EntityPlayer.getCommandSenderName().equalsIgnoreCase(((EntityTameable)this).getOwnerName()))
                 {
+                    // CraftBukkit start
+                    if (CraftEventFactory.callPlayerLeashEntityEvent(this, par1EntityPlayer, par1EntityPlayer).isCancelled())
+                    {
+                        ((EntityPlayerMP) par1EntityPlayer).playerNetServerHandler.func_147359_a(new S1BPacketEntityAttach(1, this, this.getLeashedToEntity()));
+                        return false;
+                    }
+
+                    // CraftBukkit end
                     this.setLeashedToEntity(par1EntityPlayer, true);
                     --itemstack.stackSize;
                     return true;
@@ -1097,6 +1168,7 @@
         {
             if (this.leashedToEntity == null || this.leashedToEntity.isDead)
             {
+                this.worldObj.getServer().getPluginManager().callEvent(new EntityUnleashEvent(this.getBukkitEntity(), UnleashReason.HOLDER_GONE)); // CraftBukkit
                 this.clearLeashed(true, true);
             }
         }
@@ -1186,10 +1258,16 @@
             }
             else
             {
+                this.worldObj.getServer().getPluginManager().callEvent(new EntityUnleashEvent(this.getBukkitEntity(), UnleashReason.UNKNOWN)); // CraftBukkit
                 this.clearLeashed(false, true);
             }
         }
 
         this.field_110170_bx = null;
     }
+
+    public boolean canDespawn_CB()
+    {
+        return this.canDespawn();
+    }
 }
