--- ../src-base/minecraft/net/minecraft/network/rcon/RConConsoleSource.java
+++ ../src-work/minecraft/net/minecraft/network/rcon/RConConsoleSource.java
@@ -36,6 +36,13 @@
         return new ChatComponentText(this.getCommandSenderName());
     }
 
+    // CraftBukkit start - Send a String
+    public void sendMessage(String message)
+    {
+        this.buffer.append(message);
+    }
+    // CraftBukkit end
+
     public void addChatMessage(IChatComponent ichatcomponent)
     {
         this.buffer.append(ichatcomponent.getUnformattedText());
