--- ../src-base/minecraft/net/minecraft/world/storage/SaveHandler.java
+++ ../src-work/minecraft/net/minecraft/world/storage/SaveHandler.java
@@ -10,6 +10,7 @@
 import cpw.mods.fml.common.FMLCommonHandler;
 import cpw.mods.fml.common.StartupQuery;
 import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.entity.player.EntityPlayerMP;
 import net.minecraft.nbt.CompressedStreamTools;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.server.MinecraftServer;
@@ -19,6 +20,13 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+// CraftBukkit start
+import java.util.UUID;
+
+import org.bukkit.craftbukkit.entity.CraftPlayer;
+// CraftBukkit end
+import cpw.mods.fml.common.registry.GameData; // Cauldron
+
 public class SaveHandler implements ISaveHandler, IPlayerFileData
 {
     private static final Logger logger = LogManager.getLogger();
@@ -27,6 +35,7 @@
     private final File mapDataDir;
     private final long initializationTime = MinecraftServer.getSystemTimeMillis();
     private final String saveDirectoryName;
+    private UUID uuid = null; // CraftBukkit
     private static final String __OBFID = "CL_00000585";
 
     public SaveHandler(File par1File, String par2Str, boolean par3)
@@ -65,7 +74,7 @@
         catch (IOException ioexception)
         {
             ioexception.printStackTrace();
-            throw new RuntimeException("Failed to check session lock, aborting");
+            throw new RuntimeException("Failed to check session lock for world " + this.worldDirectory + ", aborting"); // Cauldron
         }
     }
 
@@ -85,7 +94,7 @@
             {
                 if (datainputstream.readLong() != this.initializationTime)
                 {
-                    throw new MinecraftException("The save is being accessed from another location, aborting");
+                    throw new MinecraftException("The save folder for world " + this.worldDirectory + " is being accessed from another location, aborting"); // Cauldron
                 }
             }
             finally
@@ -95,7 +104,10 @@
         }
         catch (IOException ioexception)
         {
-            throw new MinecraftException("Failed to check session lock, aborting");
+            // Cauldron start
+            ioexception.printStackTrace();
+            throw new MinecraftException("Failed to check session lock for world " + this.worldDirectory + ", aborting");
+            // Cauldron end
         }
     }
 
@@ -120,6 +132,17 @@
                 nbttagcompound1 = nbttagcompound.getCompoundTag("Data");
                 worldInfo = new WorldInfo(nbttagcompound1);
                 FMLCommonHandler.instance().handleWorldDataLoad(this, worldInfo, nbttagcompound);
+                // Cauldron start - inject bukkit materials before plugins load
+                if (worldInfo.getDimension() == 0)
+                {
+                    GameData.injectBlockBukkitMaterials();
+                    GameData.injectItemBukkitMaterials();
+                    // since we modify bukkit enums, we need to guarantee that plugins are
+                    // loaded after all mods have been loaded by FML to avoid race conditions.
+                    MinecraftServer.getServer().server.loadPlugins();
+                    MinecraftServer.getServer().server.enablePlugins(org.bukkit.plugin.PluginLoadOrder.STARTUP);
+                }
+                // Cauldron end
                 return worldInfo;
             }
             catch (StartupQuery.AbortedException e)
@@ -267,6 +290,14 @@
 
         if (nbttagcompound != null)
         {
+            // CraftBukkit start
+            if (par1EntityPlayer instanceof EntityPlayerMP)
+            {
+                CraftPlayer player = (CraftPlayer) par1EntityPlayer.getBukkitEntity(); // Cauldron
+                player.setFirstPlayed(new File(playersDirectory, par1EntityPlayer.getCommandSenderName() + ".dat").lastModified());
+            }
+
+            // CraftBukkit end
             par1EntityPlayer.readFromNBT(nbttagcompound);
         }
 
@@ -323,4 +354,80 @@
     {
         return this.saveDirectoryName;
     }
+
+    // CraftBukkit start
+    public UUID getUUID()
+    {
+        if (uuid != null)
+        {
+            return uuid;
+        }
+
+        File file1 = new File(this.worldDirectory, "uid.dat");
+
+        if (file1.exists())
+        {
+            DataInputStream dis = null;
+
+            try
+            {
+                dis = new DataInputStream(new FileInputStream(file1));
+                return uuid = new UUID(dis.readLong(), dis.readLong());
+            }
+            catch (IOException ex)
+            {
+                logger.warn("Failed to read " + file1 + ", generating new random UUID", ex);
+            }
+            finally
+            {
+                if (dis != null)
+                {
+                    try
+                    {
+                        dis.close();
+                    }
+                    catch (IOException ex)
+                    {
+                        // NOOP
+                    }
+                }
+            }
+        }
+
+        uuid = UUID.randomUUID();
+        DataOutputStream dos = null;
+
+        try
+        {
+            dos = new DataOutputStream(new FileOutputStream(file1));
+            dos.writeLong(uuid.getMostSignificantBits());
+            dos.writeLong(uuid.getLeastSignificantBits());
+        }
+        catch (IOException ex)
+        {
+            logger.warn("Failed to write " + file1, ex);
+        }
+        finally
+        {
+            if (dos != null)
+            {
+                try
+                {
+                    dos.close();
+                }
+                catch (IOException ex)
+                {
+                    // NOOP
+                }
+            }
+        }
+
+        return uuid;
+    }
+
+    public File getPlayerDir()
+    {
+        return playersDirectory;
+    }
+    // CraftBukkit end
 }
