--- ../src-base/minecraft/net/minecraft/entity/item/EntityItemFrame.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityItemFrame.java
@@ -47,6 +47,15 @@
         {
             if (!this.worldObj.isRemote)
             {
+                // CraftBukkit start
+                org.bukkit.event.entity.EntityDamageEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.handleEntityDamageEvent(this, par1DamageSource, par2);
+
+                if ((event != null && event.isCancelled()) || this.isDead)
+                {
+                    return true;
+                }
+
+                // CraftBukkit end
                 this.func_146065_b(par1DamageSource.getEntity(), false);
                 this.setDisplayedItem((ItemStack)null);
             }
