--- ../src-base/minecraft/net/minecraft/client/renderer/Tessellator.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/Tessellator.java
@@ -164,8 +164,8 @@
 
             if (rawBufferSize > 0x20000 && field_147569_p < (rawBufferSize << 3))
             {
-                rawBufferSize = 0x10000;
-                rawBuffer = new int[rawBufferSize];
+                rawBufferSize = 0;
+                rawBuffer = null;
             }
 
             int i = this.field_147569_p * 4;
