--- ../src-base/minecraft/net/minecraft/client/renderer/ThreadDownloadImageData.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/ThreadDownloadImageData.java
@@ -86,23 +86,23 @@
                         httpurlconnection.setDoOutput(false);
                         httpurlconnection.connect();
 
-                        if (httpurlconnection.getResponseCode() / 100 == 2)
+                        if (httpurlconnection.getResponseCode() / 100 != 2)
                         {
-                            BufferedImage bufferedimage = ImageIO.read(httpurlconnection.getInputStream());
+                            return;
+                        }
 
-                            if (ThreadDownloadImageData.this.imageBuffer != null)
-                            {
-                                bufferedimage = ThreadDownloadImageData.this.imageBuffer.parseUserSkin(bufferedimage);
-                            }
+                        BufferedImage bufferedimage = ImageIO.read(httpurlconnection.getInputStream());
 
-                            ThreadDownloadImageData.this.func_147641_a(bufferedimage);
-                            return;
+                        if (ThreadDownloadImageData.this.imageBuffer != null)
+                        {
+                            bufferedimage = ThreadDownloadImageData.this.imageBuffer.parseUserSkin(bufferedimage);
                         }
+
+                        ThreadDownloadImageData.this.func_147641_a(bufferedimage);
                     }
                     catch (Exception exception)
                     {
                         ThreadDownloadImageData.field_147644_c.error("Couldn\'t download http texture", exception);
-                        return;
                     }
                     finally
                     {
