--- ../src-base/minecraft/org/bukkit/plugin/java/PluginClassLoader.java
+++ ../src-work/minecraft/org/bukkit/plugin/java/PluginClassLoader.java
@@ -1,5 +1,21 @@
 package org.bukkit.plugin.java;
 
+// MCPC+ start
+import net.md_5.specialsource.provider.ClassLoaderProvider;
+import net.md_5.specialsource.transformer.MavenShade;
+//import org.bouncycastle.util.io.Streams;
+import net.md_5.specialsource.*;
+import net.md_5.specialsource.repo.*;
+import org.bukkit.plugin.PluginDescriptionFile;
+import java.io.*;
+import java.net.JarURLConnection;
+import java.security.CodeSigner;
+import java.security.CodeSource;
+import java.util.concurrent.*;
+import za.co.mcportcentral.MCPCConfig;
+import za.co.mcportcentral.MCPCUtils;
+// MCPC+ end
+
 import java.net.URL;
 import java.net.URLClassLoader;
 import java.util.HashMap;
@@ -15,9 +31,33 @@
  */
 public class PluginClassLoader extends URLClassLoader {
     private final JavaPluginLoader loader;
-    private final Map<String, Class<?>> classes = new HashMap<String, Class<?>>();
+    private final ConcurrentMap<String, Class<?>> classes = new ConcurrentHashMap<String, Class<?>>(); // MCPC+ - Threadsafe classloading
     final boolean extended = this.getClass() != PluginClassLoader.class;
+    // MCPC+ start
+    private JarRemapper remapper;     // class remapper for this plugin, or null
+    private RemapperProcessor remapperProcessor; // secondary; for inheritance & remapping reflection
+    private boolean debug;            // classloader debugging
 
+    private static ConcurrentMap<Integer,JarMapping> jarMappings = new ConcurrentHashMap<Integer, JarMapping>();
+    private static final int F_USE_GUAVA10      = 1 << 1;
+    private static final int F_GLOBAL_INHERIT   = 1 << 2;
+    private static final int F_REMAP_OBCPRE     = 1 << 3;
+    private static final int F_REMAP_NMS152     = 1 << 4;
+    private static final int F_REMAP_NMS161     = 1 << 5;
+    private static final int F_REMAP_NMS162     = 1 << 6;
+    private static final int F_REMAP_NMS164     = 1 << 7;
+    private static final int F_REMAP_NMS172     = 1 << 8;
+    private static final int F_REMAP_OBC152     = 1 << 9;
+    private static final int F_REMAP_OBC161     = 1 << 10;
+    private static final int F_REMAP_OBC162     = 1 << 11;
+    private static final int F_REMAP_OBC164     = 1 << 12;
+    private static final int F_REMAP_OBC172     = 1 << 13;
+    private static final int F_REMAP_NMSPRE_MASK= 0xfff00000;  // "unversioned" NMS plugin version
+
+    // This trick bypasses Maven Shade's package rewriting when using String literals [same trick in jline]
+    private static final String org_bukkit_craftbukkit = new String(new char[] {'o','r','g','/','b','u','k','k','i','t','/','c','r','a','f','t','b','u','k','k','i','t'});
+    // MCPC+ end
+
     /**
      * Internal class not intended to be exposed
      */
@@ -35,12 +75,120 @@
         }
     }
 
-
-    PluginClassLoader(final JavaPluginLoader loader, final URL[] urls, final ClassLoader parent, final Object methodSignature) {
+    PluginClassLoader(final JavaPluginLoader loader, final URL[] urls, final ClassLoader parent, PluginDescriptionFile pluginDescriptionFile) { // MCPC+ - add PluginDescriptionFile
         super(urls, parent);
         Validate.notNull(loader, "Loader cannot be null");
 
         this.loader = loader;
+
+        // MCPC+ start
+
+        String pluginName = pluginDescriptionFile.getName();
+
+        // configure default remapper settings
+        boolean useCustomClassLoader = MCPCConfig.getBoolean("plugin-settings.default.custom-class-loader", true);
+        debug = MCPCConfig.getBoolean("plugin-settings.default.debug", false);
+        boolean useGuava10 = MCPCConfig.getBoolean("plugin-settings.default.use-guava10", true);
+        boolean remapNMS172 = MCPCConfig.getBoolean("plugin-settings.default.remap-nms-v1_7_R1", true);
+        boolean remapNMS164 = MCPCConfig.getBoolean("plugin-settings.default.remap-nms-v1_6_R3", true);
+        boolean remapNMS162 = MCPCConfig.getBoolean("plugin-settings.default.remap-nms-v1_6_R2", true);
+        boolean remapNMS161 = MCPCConfig.getBoolean("plugin-settings.default.remap-nms-v1_6_R1", true);
+        boolean remapNMS152 = MCPCConfig.getBoolean("plugin-settings.default.remap-nms-v1_5_R3", true);
+        String remapNMSPre = MCPCConfig.getString("plugin-settings.default.remap-nms-pre", "false");
+        boolean remapOBC172 = MCPCConfig.getBoolean("plugin-settings.default.remap-obc-v1_7_R1", false);
+        boolean remapOBC164 = MCPCConfig.getBoolean("plugin-settings.default.remap-obc-v1_6_R3", false);
+        boolean remapOBC162 = MCPCConfig.getBoolean("plugin-settings.default.remap-obc-v1_6_R2", false);
+        boolean remapOBC161 = MCPCConfig.getBoolean("plugin-settings.default.remap-obc-v1_6_R1", false);
+        boolean remapOBC152 = MCPCConfig.getBoolean("plugin-settings.default.remap-obc-v1_5_R3", true);
+        boolean remapOBCPre = MCPCConfig.getBoolean("plugin-settings.default.remap-obc-pre", false);
+        boolean globalInherit = MCPCConfig.getBoolean("plugin-settings.default.global-inheritance", true);
+        boolean pluginInherit = MCPCConfig.getBoolean("plugin-settings.default.plugin-inheritance", true);
+        boolean reflectFields = MCPCConfig.getBoolean("plugin-settings.default.remap-reflect-field", true);
+        boolean reflectClass = MCPCConfig.getBoolean("plugin-settings.default.remap-reflect-class", true);
+        boolean allowFuture = MCPCConfig.getBoolean("plugin-settings.default.remap-allow-future", false);
+
+        // plugin-specific overrides
+        useCustomClassLoader = MCPCConfig.getBoolean("plugin-settings."+pluginName+".custom-class-loader", useCustomClassLoader, false);
+        debug = MCPCConfig.getBoolean("plugin-settings."+pluginName+".debug", debug, false);
+        useGuava10 = MCPCConfig.getBoolean("plugin-settings."+pluginName+".use-guava10", useGuava10, false);
+        remapNMS172 = MCPCConfig.getBoolean("plugin-settings."+pluginName+".remap-nms-v1_7_R1", remapNMS172, false);
+        remapNMS164 = MCPCConfig.getBoolean("plugin-settings."+pluginName+".remap-nms-v1_6_R3", remapNMS164, false);
+        remapNMS162 = MCPCConfig.getBoolean("plugin-settings."+pluginName+".remap-nms-v1_6_R2", remapNMS162, false);
+        remapNMS161 = MCPCConfig.getBoolean("plugin-settings."+pluginName+".remap-nms-v1_6_R1", remapNMS161, false);
+        remapNMS152 = MCPCConfig.getBoolean("plugin-settings."+pluginName+".remap-nms-v1_5_R3", remapNMS152, false);
+        remapNMSPre = MCPCConfig.getString("plugin-settings."+pluginName+".remap-nms-pre", remapNMSPre, false);
+        remapOBC164 = MCPCConfig.getBoolean("plugin-settings."+pluginName+".remap-obc-v1_6_R3", remapOBC164, false);
+        remapOBC162 = MCPCConfig.getBoolean("plugin-settings."+pluginName+".remap-obc-v1_6_R2", remapOBC162, false);
+        remapOBC161 = MCPCConfig.getBoolean("plugin-settings."+pluginName+".remap-obc-v1_6_R1", remapOBC161, false);
+        remapOBC152 = MCPCConfig.getBoolean("plugin-settings."+pluginName+".remap-obc-v1_5_R3", remapOBC152, false);
+        remapOBCPre = MCPCConfig.getBoolean("plugin-settings."+pluginName+".remap-obc-pre", remapOBCPre, false);
+        globalInherit = MCPCConfig.getBoolean("plugin-settings."+pluginName+".global-inheritance", globalInherit, false);
+        pluginInherit = MCPCConfig.getBoolean("plugin-settings."+pluginName+".plugin-inheritance", pluginInherit, false);
+        reflectFields = MCPCConfig.getBoolean("plugin-settings."+pluginName+".remap-reflect-field", reflectFields, false);
+        reflectClass = MCPCConfig.getBoolean("plugin-settings."+pluginName+".remap-reflect-class", reflectClass, false);
+        allowFuture = MCPCConfig.getBoolean("plugin-settings."+pluginName+".remap-allow-future", allowFuture, false);
+
+        if (debug) {
+            System.out.println("PluginClassLoader debugging enabled for "+pluginName);
+        }
+
+        if (!useCustomClassLoader) {
+            remapper = null;
+            return;
+        }
+
+        int flags = 0;
+        if (useGuava10) flags |= F_USE_GUAVA10;
+        if (remapNMS172) flags |= F_REMAP_NMS172;
+        if (remapNMS164) flags |= F_REMAP_NMS164;
+        if (remapNMS162) flags |= F_REMAP_NMS162;
+        if (remapNMS161) flags |= F_REMAP_NMS161;
+        if (remapNMS152) flags |= F_REMAP_NMS152;
+        //if (remapNMS146) flags |= F_REMAP_NMS146;
+        if (!remapNMSPre.equals("false")) {
+            if      (remapNMSPre.equals("1.7.2")) flags |= 0x17200000;
+            else if (remapNMSPre.equals("1.6.4")) flags |= 0x16400000;
+            else if (remapNMSPre.equals("1.6.2")) flags |= 0x16200000;
+            else if (remapNMSPre.equals("1.6.1")) flags |= 0x16100000;
+            else if (remapNMSPre.equals("1.5.2")) flags |= 0x15200000;
+            else {
+                System.out.println("Unsupported nms-remap-pre version '"+remapNMSPre+"', disabling");
+            }
+        }
+        if (remapOBC172) flags |= F_REMAP_OBC172;
+        if (remapOBC164) flags |= F_REMAP_OBC164;
+        if (remapOBC162) flags |= F_REMAP_OBC162;
+        if (remapOBC161) flags |= F_REMAP_OBC161;
+        if (remapOBC152) flags |= F_REMAP_OBC152;
+        if (remapOBCPre) flags |= F_REMAP_OBCPRE;
+        if (globalInherit) flags |= F_GLOBAL_INHERIT;
+
+        JarMapping jarMapping = getJarMapping(flags);
+
+        // Load inheritance map
+        if ((flags & F_GLOBAL_INHERIT) != 0) {
+            if (debug) {
+                System.out.println("Enabling global inheritance remapping");
+                //ClassLoaderProvider.verbose = debug; // TODO: changed in https://github.com/md-5/SpecialSource/commit/132584eda4f0860c9d14f4c142e684a027a128b8#L3L48
+            }
+            jarMapping.setInheritanceMap(loader.getGlobalInheritanceMap());
+            jarMapping.setFallbackInheritanceProvider(new ClassLoaderProvider(this));
+        }
+
+        remapper = new JarRemapper(jarMapping);
+
+        if (pluginInherit || reflectFields || reflectClass) {
+            remapperProcessor = new RemapperProcessor(
+                    pluginInherit ? loader.getGlobalInheritanceMap() : null,
+                    (reflectFields || reflectClass) ? jarMapping : null);
+
+            remapperProcessor.setRemapReflectField(reflectFields);
+            remapperProcessor.setRemapReflectClass(reflectClass);
+            remapperProcessor.debug = debug;
+        } else {
+            remapperProcessor = null;
+        }
+        // MCPC+ end
     }
 
     @Override
@@ -65,48 +213,266 @@
         return findClass0(name, checkGlobal);
     }
 
+    /**
+     * @deprecated Internal method that wasn't intended to be exposed
+     */
+    @Deprecated
+    public Set<String> getClasses() {
+        if (loader.warn) {
+            loader.server.getLogger().log(Level.WARNING, "Method \"public Set<String> getClasses()\" is Deprecated, and may be removed in a future version of Bukkit", new AuthorNagException(""));
+            loader.warn = false;
+        }
+        return getClasses0();
+    }
+
+    Set<String> getClasses0() {
+        return classes.keySet();
+    }
+
+    // MCPC+ start
+    /**
+     * Get the "native" obfuscation version, from our Maven shading version.
+     */
+    public static String getNativeVersion() {
+        // see https://github.com/mbax/VanishNoPacket/blob/master/src/main/java/org/kitteh/vanish/compat/NMSManager.java
+        final String packageName = org.bukkit.craftbukkit.CraftServer.class.getPackage().getName();
+        //if (MCPCUtils.deobfuscatedEnvironment())
+            return "v1_7_R1";
+        //else return packageName.substring(packageName.lastIndexOf('.')  + 1);
+    }
+
+    /**
+     * Load NMS mappings from CraftBukkit mc-dev to repackaged srgnames for FML runtime deobf
+     *
+     * @param jarMapping An existing JarMappings instance to load into
+     * @param obfVersion CraftBukkit version with internal obfuscation counter identifier
+     *                   >=1.4.7 this is the major version + R#. v1_4_R1=1.4.7, v1_5_R1=1.5, v1_5_R2=1.5.1..
+     *                   For older versions (including pre-safeguard) it is the full Minecraft version number
+     * @throws IOException
+     */
+    private void loadNmsMappings(JarMapping jarMapping, String obfVersion) throws IOException {
+        Map<String, String> relocations = new HashMap<String, String>();
+        // mc-dev jar to CB, apply version shading (aka plugin safeguard)
+        relocations.put("net.minecraft.server", "net.minecraft.server." + obfVersion);
+
+        jarMapping.loadMappings(
+                new BufferedReader(new InputStreamReader(loader.getClass().getClassLoader().getResourceAsStream("mappings/"+obfVersion+"/cb2numpkg.srg"))),
+                new MavenShade(relocations),
+                null, false);
+
+        // resolve naming conflict in FML/CB
+        jarMapping.methods.put("net/minecraft/server/"+obfVersion+"/PlayerConnection/getPlayer ()Lorg/bukkit/craftbukkit/"+getNativeVersion()+"/entity/CraftPlayer;", "getPlayerB");
+
+        // remap bouncycastle to Forge's included copy, not the vanilla obfuscated copy (not in MCPC+), see #133
+        //jarMapping.packages.put("net/minecraft/"+obfVersion+"/org/bouncycastle", "org/bouncycastle"); No longer needed
+    }
+
+    private JarMapping getJarMapping(int flags) {
+        JarMapping jarMapping = jarMappings.get(flags);
+
+        if (jarMapping != null) {
+            if (debug) {
+                System.out.println("Mapping reused for "+Integer.toHexString(flags));
+            }
+            return jarMapping;
+        }
+
+        jarMapping = new JarMapping();
+        try {
+
+            if ((flags & F_USE_GUAVA10) != 0) {
+                // Guava 10 is part of the Bukkit API, so plugins can use it, but FML includes Guava 15
+                // To resolve this conflict, remap plugin usages to Guava 10 in a separate package
+                // Most plugins should keep this enabled, unless they want a newer Guava
+                jarMapping.packages.put("com/google/common", "guava10/com/google/common");
+            }
+            jarMapping.packages.put(org_bukkit_craftbukkit + "/libs/com/google/gson", "com/google/gson"); // Handle Gson being in a "normal" place
+
+            if ((flags & F_REMAP_NMS172) != 0) {
+                loadNmsMappings(jarMapping, "v1_7_R1");
+            }
+
+            if ((flags & F_REMAP_NMS164) != 0) {
+                loadNmsMappings(jarMapping, "v1_6_R3");
+            }
+
+            if ((flags & F_REMAP_NMS162) != 0) {
+                loadNmsMappings(jarMapping, "v1_6_R2");
+            }
+
+            if ((flags & F_REMAP_NMS161) != 0) {
+                loadNmsMappings(jarMapping, "v1_6_R1");
+            }
+
+            if ((flags & F_REMAP_NMS152) != 0) {
+                loadNmsMappings(jarMapping, "v1_5_R3");
+            }
+
+            if ((flags & F_REMAP_OBC172) != 0) {
+                jarMapping.packages.put(org_bukkit_craftbukkit+"/v1_7_R1", org_bukkit_craftbukkit+"/"+getNativeVersion());
+            }
+
+            if ((flags & F_REMAP_OBC164) != 0) {
+                jarMapping.packages.put(org_bukkit_craftbukkit+"/v1_6_R3", org_bukkit_craftbukkit+"/"+getNativeVersion());
+            }
+
+            if ((flags & F_REMAP_OBC162) != 0) {
+                jarMapping.packages.put(org_bukkit_craftbukkit+"/v1_6_R2", org_bukkit_craftbukkit+"/"+getNativeVersion());
+            }
+
+            if ((flags & F_REMAP_OBC161) != 0) {
+                jarMapping.packages.put(org_bukkit_craftbukkit+"/v1_6_R1", org_bukkit_craftbukkit+"/"+getNativeVersion());
+            }
+
+            if ((flags & F_REMAP_OBC152) != 0) {
+                jarMapping.packages.put(org_bukkit_craftbukkit+"/v1_5_R3", org_bukkit_craftbukkit+"/"+getNativeVersion());
+            }
+
+            if ((flags & F_REMAP_OBCPRE) != 0) {
+                // enabling unversioned obc not currently compatible with versioned obc plugins (overmapped) -
+                // admins should enable remap-obc-pre on a per-plugin basis, as needed
+                //jarMapping.packages.put(org_bukkit_craftbukkit+"/v1_4_R1", org_bukkit_craftbukkit+"/v1_4_R1");
+
+                // then map unversioned to current version
+                jarMapping.packages.put(org_bukkit_craftbukkit+"/libs/org/objectweb/asm", "org/objectweb/asm"); // ?
+                jarMapping.packages.put(org_bukkit_craftbukkit, org_bukkit_craftbukkit+"/"+getNativeVersion());
+            }
+
+            if ((flags & F_REMAP_NMSPRE_MASK) != 0) {
+                String filename;
+                switch (flags & F_REMAP_NMSPRE_MASK)
+                {
+                    case 0x01720000: filename = "mappings/v1_7_R1/cb2numpkg.srg"; break;
+                    case 0x01640000: filename = "mappings/v1_6_R3/cb2numpkg.srg"; break;
+                    case 0x01620000: filename = "mappings/v1_6_R2/cb2numpkg.srg"; break;
+                    case 0x01610000: filename = "mappings/v1_6_R1/cb2numpkg.srg"; break;
+                    case 0x01510000: filename = "mappings/v1_5_R2/cb2numpkg.srg"; break;
+                    default: throw new IllegalArgumentException("Invalid unversioned mapping flags: "+Integer.toHexString(flags & F_REMAP_NMSPRE_MASK)+" in "+Integer.toHexString(flags));
+                }
+
+                jarMapping.loadMappings(
+                        new BufferedReader(new InputStreamReader(loader.getClass().getClassLoader().getResourceAsStream(filename))),
+                        null, // no version relocation!
+                        null, false);
+            }
+
+            System.out.println("Mapping loaded "+jarMapping.packages.size()+" packages, "+jarMapping.classes.size()+" classes, "+jarMapping.fields.size()+" fields, "+jarMapping.methods.size()+" methods, flags "+Integer.toHexString(flags));
+
+            JarMapping currentJarMapping = jarMappings.putIfAbsent(flags, jarMapping);
+            return currentJarMapping == null ? jarMapping : currentJarMapping;
+        } catch (IOException ex) {
+            ex.printStackTrace();
+            throw new RuntimeException(ex);
+        }
+    }
+
     Class<?> findClass0(String name, boolean checkGlobal) throws ClassNotFoundException {
         if (name.startsWith("org.bukkit.") || name.startsWith("net.minecraft.")) {
+            if (debug) {
+                System.out.println("Unexpected plugin findClass on OBC/NMS: name="+name+", checkGlobal="+checkGlobal+"; returning not found");
+            }
             throw new ClassNotFoundException(name);
         }
-        Class<?> result = classes.get(name);
+        // MCPC+ start - custom loader, if enabled, threadsafety
+        synchronized (name.intern()) {
+            Class<?> result = classes.get(name);
 
-        if (result == null) {
-            if (checkGlobal) {
-                result = loader.extended ? loader.getClassByName(name) : loader.getClassByName0(name); // Don't warn on deprecation, but maintain overridability
-            }
-
             if (result == null) {
-                result = super.findClass(name);
+                if (checkGlobal) {
+                    result = loader.extended ? loader.getClassByName(name) : loader.getClassByName0(name); // Don't warn on deprecation, but maintain overridability
+                }
 
-                if (result != null) {
-                    if (loader.extended) { // Don't warn on deprecation, but maintain overridability
-                        loader.setClass(name, result);
+                if (result == null) {
+                    if (remapper == null) {
+                        result = super.findClass(name);
                     } else {
-                        loader.setClass0(name, result);
+                        result = remappedFindClass(name);
                     }
+
+                    if (result != null) {
+                        if (loader.extended) { // Don't warn on deprecation, but maintain overridability
+                            loader.setClass(name, result);
+                        } else {
+                            loader.setClass0(name, result);
+                        }
+                    }
                 }
+                if (result != null) {
+                    Class<?> old = classes.putIfAbsent(name, result);
+                    if (old != null && old != result) {
+                        System.err.println("Defined class " + name + " twice as different classes, " + result + " and " + old);
+                        result = old;
+                    }
+                }
             }
 
-            classes.put(name, result);
+            return result;
         }
-
-        return result;
+        // MCPC+ end
     }
+    private Class<?> remappedFindClass(String name) throws ClassNotFoundException {
+        Class<?> result = null;
 
-    /**
-     * @deprecated Internal method that wasn't intended to be exposed
-     */
-    @Deprecated
-    public Set<String> getClasses() {
-        if (loader.warn) {
-            loader.server.getLogger().log(Level.WARNING, "Method \"public Set<String> getClasses()\" is Deprecated, and may be removed in a future version of Bukkit", new AuthorNagException(""));
-            loader.warn = false;
+        try {
+            // Load the resource to the name
+            String path = name.replace('.', '/').concat(".class");
+            URL url = this.findResource(path);
+            if (url != null) {
+                InputStream stream = url.openStream();
+                if (stream != null) {
+                    byte[] bytecode = null;
+
+                    // Reflection remap and inheritance extract
+                    if (remapperProcessor != null) {
+                        // add to inheritance map
+                        bytecode = remapperProcessor.process(stream);
+                        if (bytecode == null) stream = url.openStream();
+                    }
+
+                    /*if (bytecode == null) {
+                        bytecode = Streams.readAll(stream);
+                    }*/
+
+                    // Remap the classes
+                    byte[] remappedBytecode = remapper.remapClassFile(bytecode, RuntimeRepo.getInstance());
+
+                    if (debug) {
+                        File file = new File("remapped-plugin-classes/"+name+".class");
+                        file.getParentFile().mkdirs();
+                        try {
+                            FileOutputStream fileOutputStream = new FileOutputStream(file);
+                            fileOutputStream.write(remappedBytecode);
+                            fileOutputStream.close();
+                        } catch (IOException ex) {
+                            ex.printStackTrace();
+                        }
+                    }
+
+                    // Define (create) the class using the modified byte code
+                    // The top-child class loader is used for this to prevent access violations
+                    // Set the codesource to the jar, not within the jar, for compatibility with
+                    // plugins that do new File(getClass().getProtectionDomain().getCodeSource().getLocation().toURI()))
+                    // instead of using getResourceAsStream - see https://github.com/MinecraftPortCentral/MCPC-Plus/issues/75
+                    JarURLConnection jarURLConnection = (JarURLConnection) url.openConnection(); // parses only
+                    URL jarURL = jarURLConnection.getJarFileURL();
+                    CodeSource codeSource = new CodeSource(jarURL, new CodeSigner[0]);
+
+                    result = this.defineClass(name, remappedBytecode, 0, remappedBytecode.length, codeSource);
+                    if (result != null) {
+                        // Resolve it - sets the class loader of the class
+                        this.resolveClass(result);
+                    }
+                }
+            }
+        } catch (Throwable t) {
+            if (debug) {
+                System.out.println("remappedFindClass("+name+") exception: "+t);
+                t.printStackTrace();
+            }
+            throw new ClassNotFoundException("Failed to remap class "+name, t);
         }
-        return getClasses0();
-    }
 
-    Set<String> getClasses0() {
-        return classes.keySet();
+        return result;
     }
+    // MCPC+ end
 }
