--- ../src-base/minecraft/net/minecraft/world/WorldManager.java
+++ ../src-work/minecraft/net/minecraft/world/WorldManager.java
@@ -14,7 +14,7 @@
     // JAVADOC FIELD $$ field_72783_a
     private MinecraftServer mcServer;
     // JAVADOC FIELD $$ field_72782_b
-    private WorldServer theWorldServer;
+    public WorldServer theWorldServer; // CraftBukkit - private -> public
     private static final String __OBFID = "CL_00001433";
 
     public WorldManager(MinecraftServer par1MinecraftServer, WorldServer par2WorldServer)
