--- ../src-base/minecraft/net/minecraft/entity/monster/EntityWitch.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntityWitch.java
@@ -214,6 +214,8 @@
     // JAVADOC METHOD $$ func_70628_a
     protected void dropFewItems(boolean par1, int par2)
     {
+        // CraftBukkit start
+        java.util.List<org.bukkit.inventory.ItemStack> loot = new java.util.ArrayList<org.bukkit.inventory.ItemStack>();
         int j = this.rand.nextInt(3) + 1;
 
         for (int k = 0; k < j; ++k)
@@ -226,11 +228,14 @@
                 l += this.rand.nextInt(par2 + 1);
             }
 
-            for (int i1 = 0; i1 < l; ++i1)
+            if (l > 0)
             {
-                this.func_145779_a(item, 1);
+                loot.add(new org.bukkit.inventory.ItemStack(org.bukkit.craftbukkit.util.CraftMagicNumbers.getMaterial(item), l));
             }
         }
+
+        org.bukkit.craftbukkit.event.CraftEventFactory.callEntityDeathEvent(this, loot);
+        // CraftBukkit end
     }
 
     // JAVADOC METHOD $$ func_82196_d
