--- ../src-base/minecraft/net/minecraft/entity/monster/EntitySkeleton.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntitySkeleton.java
@@ -36,6 +36,8 @@
 import net.minecraft.world.World;
 import net.minecraft.world.WorldProviderHell;
 
+import org.bukkit.event.entity.EntityCombustEvent; // CraftBukkit
+
 public class EntitySkeleton extends EntityMob implements IRangedAttackMob
 {
     private EntityAIArrowAttack aiArrowAttack = new EntityAIArrowAttack(this, 1.0D, 20, 60, 15.0F);
@@ -154,7 +156,16 @@
 
                 if (flag)
                 {
-                    this.setFire(8);
+                    // CraftBukkit start
+                    EntityCombustEvent event = new EntityCombustEvent(this.getBukkitEntity(), 8);
+                    this.worldObj.getServer().getPluginManager().callEvent(event);
+
+                    if (!event.isCancelled())
+                    {
+                        this.setFire(event.getDuration());
+                    }
+
+                    // CraftBukkit end
                 }
             }
         }
@@ -205,42 +216,71 @@
     // JAVADOC METHOD $$ func_70628_a
     protected void dropFewItems(boolean par1, int par2)
     {
-        int j;
-        int k;
+        // CraftBukkit start - Whole method
+        java.util.List<org.bukkit.inventory.ItemStack> loot = new java.util.ArrayList<org.bukkit.inventory.ItemStack>();
 
         if (this.getSkeletonType() == 1)
         {
-            j = this.rand.nextInt(3 + par2) - 1;
+            int count = this.rand.nextInt(3 + par2) - 1;
 
-            for (k = 0; k < j; ++k)
+            if (count > 0)
             {
-                this.func_145779_a(Items.coal, 1);
+                loot.add(new org.bukkit.inventory.ItemStack(org.bukkit.Material.COAL, count));
             }
         }
         else
         {
-            j = this.rand.nextInt(3 + par2);
+            int count = this.rand.nextInt(3 + par2);
 
-            for (k = 0; k < j; ++k)
+            if (count > 0)
             {
-                this.func_145779_a(Items.arrow, 1);
+                loot.add(new org.bukkit.inventory.ItemStack(org.bukkit.Material.ARROW, count));
             }
         }
 
-        j = this.rand.nextInt(3 + par2);
+        int count = this.rand.nextInt(3 + par2);
 
-        for (k = 0; k < j; ++k)
+        if (count > 0)
         {
-            this.func_145779_a(Items.bone, 1);
+            loot.add(new org.bukkit.inventory.ItemStack(org.bukkit.Material.BONE, count));
         }
+
+        // Determine rare item drops and add them to the loot
+        if (this.recentlyHit > 0)
+        {
+            int k = this.rand.nextInt(200) - par2;
+
+            if (k < 5)
+            {
+                ItemStack itemstack = this.getRareDrop(k <= 0 ? 1 : 0);
+
+                if (itemstack != null)
+                {
+                    loot.add(org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemstack));
+                }
+            }
+        }
+
+        org.bukkit.craftbukkit.event.CraftEventFactory.callEntityDeathEvent(this, loot);
+        // CraftBukkit end
     }
 
+    // MCPC+ start - vanilla compatibility
     protected void dropRareDrop(int par1)
     {
+        this.getRareDrop(par1);
+    }
+    // MCPC+ end
+
+    // CraftBukkit - return rare dropped item instead of dropping it
+    protected ItemStack getRareDrop(int i)
+    {
         if (this.getSkeletonType() == 1)
         {
-            this.entityDropItem(new ItemStack(Items.skull, 1, 1), 0.0F);
+            return new ItemStack(Items.skull, 1, 1); // CraftBukkit
         }
+
+        return null; // CraftBukkit
     }
 
     // JAVADOC METHOD $$ func_82164_bB
