--- ../src-base/minecraft/net/minecraft/entity/monster/EntityMagmaCube.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntityMagmaCube.java
@@ -62,6 +62,8 @@
 
     protected void dropFewItems(boolean par1, int par2)
     {
+        // CraftBukkit start - Whole method
+        java.util.List<org.bukkit.inventory.ItemStack> loot = new java.util.ArrayList<org.bukkit.inventory.ItemStack>();
         Item item = this.getDropItem();
 
         if (item != null && this.getSlimeSize() > 1)
@@ -73,11 +75,14 @@
                 j += this.rand.nextInt(par2 + 1);
             }
 
-            for (int k = 0; k < j; ++k)
+            if (j > 0)
             {
-                this.dropItem(item, 1);
+                loot.add(new org.bukkit.inventory.ItemStack(org.bukkit.craftbukkit.util.CraftMagicNumbers.getMaterial(item), j));
             }
         }
+
+        org.bukkit.craftbukkit.event.CraftEventFactory.callEntityDeathEvent(this, loot);
+        // CraftBukkit end
     }
 
     public boolean isBurning()
