--- ../src-base/minecraft/net/minecraft/entity/monster/EntityIronGolem.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntityIronGolem.java
@@ -2,6 +2,7 @@
 
 import cpw.mods.fml.relauncher.Side;
 import cpw.mods.fml.relauncher.SideOnly;
+import org.bukkit.craftbukkit.inventory.CraftItemStack; // CraftBukkit
 import net.minecraft.block.Block;
 import net.minecraft.block.material.Material;
 import net.minecraft.entity.Entity;
@@ -233,20 +234,25 @@
     // JAVADOC METHOD $$ func_70628_a
     protected void dropFewItems(boolean par1, int par2)
     {
+        // CraftBukkit start
+        java.util.List<org.bukkit.inventory.ItemStack> loot = new java.util.ArrayList<org.bukkit.inventory.ItemStack>();
         int j = this.rand.nextInt(3);
         int k;
 
-        for (k = 0; k < j; ++k)
+        if (j > 0)
         {
-            this.func_145778_a(Item.func_150898_a(Blocks.red_flower), 1, 0.0F);
+            loot.add(CraftItemStack.asNewCraftStack(Item.func_150898_a(Blocks.red_flower), j));
         }
 
         k = 3 + this.rand.nextInt(3);
 
-        for (int l = 0; l < k; ++l)
+        if (k > 0)
         {
-            this.func_145779_a(Items.iron_ingot, 1);
+            loot.add(CraftItemStack.asNewCraftStack(Items.iron_ingot, k));
         }
+
+        org.bukkit.craftbukkit.event.CraftEventFactory.callEntityDeathEvent(this, loot);
+        // CraftBukkit end
     }
 
     public int getHoldRoseTick()
