--- ../src-base/minecraft/net/minecraft/enchantment/Enchantment.java
+++ ../src-work/minecraft/net/minecraft/enchantment/Enchantment.java
@@ -50,12 +50,14 @@
 
         if (enchantmentsList[p_i1926_1_] != null)
         {
-            throw new IllegalArgumentException("Duplicate enchantment id!");
+            throw new IllegalArgumentException("Duplicate enchantment id! " + this.getClass() + " and " + enchantmentsList[p_i1926_1_].getClass() + " Enchantment ID:" + p_i1926_1_);
         }
         else
         {
             enchantmentsList[p_i1926_1_] = this;
         }
+
+        org.bukkit.enchantments.Enchantment.registerEnchantment(new org.bukkit.craftbukkit.enchantments.CraftEnchantment(this)); // CraftBukkit
     }
 
     public int getWeight()
