--- ../src-base/minecraft/net/minecraft/block/BlockCactus.java
+++ ../src-work/minecraft/net/minecraft/block/BlockCactus.java
@@ -16,6 +16,10 @@
 import net.minecraftforge.common.EnumPlantType;
 import net.minecraftforge.common.util.ForgeDirection;
 import net.minecraftforge.common.IPlantable;
+// CraftBukkit start
+import net.minecraft.entity.EntityLivingBase;
+import org.bukkit.event.entity.EntityDamageByBlockEvent;
+// CraftBukkit end
 
 public class BlockCactus extends Block implements IPlantable
 {
@@ -48,9 +52,9 @@
             {
                 int i1 = p_149674_1_.getBlockMetadata(p_149674_2_, p_149674_3_, p_149674_4_);
 
-                if (i1 == 15)
+                if (i1 >= (byte) range(3, (p_149674_1_.growthOdds / p_149674_1_.spigotConfig.cactusModifier * 15) + 0.5F, 15))   // Spigot
                 {
-                    p_149674_1_.setBlock(p_149674_2_, p_149674_3_ + 1, p_149674_4_, this);
+                    org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(p_149674_1_, p_149674_2_, p_149674_3_ + 1, p_149674_4_, this, 0); // CraftBukkit
                     p_149674_1_.setBlockMetadataWithNotify(p_149674_2_, p_149674_3_, p_149674_4_, 0, 4);
                     this.onNeighborBlockChange(p_149674_1_, p_149674_2_, p_149674_3_ + 1, p_149674_4_, this);
                 }
@@ -146,6 +150,24 @@
     // JAVADOC METHOD $$ func_149670_a
     public void onEntityCollidedWithBlock(World p_149670_1_, int p_149670_2_, int p_149670_3_, int p_149670_4_, Entity p_149670_5_)
     {
+        // CraftBukkit start - EntityDamageByBlock event
+        if (p_149670_5_ instanceof EntityLivingBase)
+        {
+            org.bukkit.block.Block damager = p_149670_1_.getWorld().getBlockAt(p_149670_2_, p_149670_3_, p_149670_4_);
+            org.bukkit.entity.Entity damagee = (p_149670_5_ == null) ? null : p_149670_5_.getBukkitEntity();
+            EntityDamageByBlockEvent event = new EntityDamageByBlockEvent(damager, damagee, org.bukkit.event.entity.EntityDamageEvent.DamageCause.CONTACT, 1D);
+            p_149670_1_.getServer().getPluginManager().callEvent(event);
+
+            if (!event.isCancelled())
+            {
+                damagee.setLastDamageCause(event);
+                p_149670_5_.attackEntityFrom(DamageSource.cactus, (float) event.getDamage());
+            }
+
+            return;
+        }
+
+        // CraftBukkit end
         p_149670_5_.attackEntityFrom(DamageSource.cactus, 1.0F);
     }
 
