--- ../src-base/minecraft/net/minecraft/block/BlockFlowerPot.java
+++ ../src-work/minecraft/net/minecraft/block/BlockFlowerPot.java
@@ -144,6 +144,7 @@
         if (tileentityflowerpot != null && tileentityflowerpot.func_145965_a() != null)
         {
             this.func_149642_a(p_149690_1_, p_149690_2_, p_149690_3_, p_149690_4_, new ItemStack(tileentityflowerpot.func_145965_a(), 1, tileentityflowerpot.func_145966_b()));
+            tileentityflowerpot.func_145964_a(null, 0);   // Spigot
         }
     }
 
@@ -154,6 +155,7 @@
         if (tileentityflowerpot != null && tileentityflowerpot.func_145965_a() != null)
         {
             this.func_149642_a(p_149749_1_, p_149749_2_, p_149749_3_, p_149749_4_, new ItemStack(tileentityflowerpot.func_145965_a(), 1, tileentityflowerpot.func_145966_b()));
+            tileentityflowerpot.func_145964_a(null, 0);   // Spigot
         }
 
         super.func_149749_a(p_149749_1_, p_149749_2_, p_149749_3_, p_149749_4_, p_149749_5_, p_149749_6_);
