--- ../src-base/minecraft/net/minecraft/block/BlockPane.java
+++ ../src-work/minecraft/net/minecraft/block/BlockPane.java
@@ -191,7 +191,7 @@
 
     public boolean canPaneConnectTo(IBlockAccess world, int x, int y, int z, ForgeDirection dir)
     {
-        return func_150098_a(world.func_147439_a(x, y, z)) || 
-                world.isSideSolid(x, y, z, dir.getOpposite(), false);
+        return func_150098_a(world.func_147439_a(x + dir.offsetX, y + dir.offsetY, z + dir.offsetZ)) || 
+                world.isSideSolid(x + dir.offsetX, y + dir.offsetY, z + dir.offsetZ, dir.getOpposite(), false);
     }
 }
