--- ../src-base/minecraft/net/minecraft/potion/Potion.java
+++ ../src-work/minecraft/net/minecraft/potion/Potion.java
@@ -17,6 +17,12 @@
 import net.minecraft.util.DamageSource;
 import net.minecraft.util.StringUtils;
 
+// CraftBukkit start
+import net.minecraft.entity.projectile.EntityPotion;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason;
+// CraftBukkit end
+
 public class Potion
 {
     // JAVADOC FIELD $$ field_76425_a
@@ -94,6 +100,7 @@
         }
 
         this.liquidColor = par3;
+        org.bukkit.potion.PotionEffectType.registerPotionEffectType(new org.bukkit.craftbukkit.potion.CraftPotionEffectType(this)); // CraftBukkit
     }
 
     // JAVADOC METHOD $$ func_76399_b
@@ -115,14 +122,14 @@
         {
             if (par1EntityLivingBase.getHealth() < par1EntityLivingBase.getMaxHealth())
             {
-                par1EntityLivingBase.heal(1.0F);
+                par1EntityLivingBase.heal(1.0F, RegainReason.MAGIC_REGEN); // CraftBukkit
             }
         }
         else if (this.id == poison.id)
         {
             if (par1EntityLivingBase.getHealth() > 1.0F)
             {
-                par1EntityLivingBase.attackEntityFrom(DamageSource.magic, 1.0F);
+                par1EntityLivingBase.attackEntityFrom(CraftEventFactory.POISON, 1.0F); // CraftBukkit - DamageSource.MAGIC -> CraftEventFactory.POISON
             }
         }
         else if (this.id == wither.id)
@@ -149,13 +156,20 @@
         }
         else
         {
-            par1EntityLivingBase.heal((float)Math.max(4 << par2, 0));
+            par1EntityLivingBase.heal((float) Math.max(4 << par2, 0), RegainReason.MAGIC); // CraftBukkit
         }
     }
 
     // JAVADOC METHOD $$ func_76402_a
     public void affectEntity(EntityLivingBase par1EntityLivingBase, EntityLivingBase par2EntityLivingBase, int par3, double par4)
     {
+        // CraftBukkit start - Delegate; we need EntityPotion
+        applyInstantEffect(par1EntityLivingBase, par2EntityLivingBase, par3, par4, null);
+    }
+
+    public void applyInstantEffect(EntityLivingBase par1EntityLivingBase, EntityLivingBase par2EntityLivingBase, int par3, double par4, EntityPotion potion)
+    {
+        // CraftBukkit end
         int j;
 
         if ((this.id != heal.id || par2EntityLivingBase.isEntityUndead()) && (this.id != harm.id || !par2EntityLivingBase.isEntityUndead()))
@@ -170,14 +184,15 @@
                 }
                 else
                 {
-                    par2EntityLivingBase.attackEntityFrom(DamageSource.causeIndirectMagicDamage(par2EntityLivingBase, par1EntityLivingBase), (float)j);
+                    // CraftBukkit - The "damager" needs to be the potion
+                    par2EntityLivingBase.attackEntityFrom(DamageSource.causeIndirectMagicDamage(potion != null ? potion : par2EntityLivingBase, par1EntityLivingBase), (float)j);
                 }
             }
         }
         else
         {
             j = (int)(par4 * (double)(4 << par3) + 0.5D);
-            par2EntityLivingBase.heal((float)j);
+            par2EntityLivingBase.heal((float)j, RegainReason.MAGIC); // CraftBukkit
         }
     }
 
