--- ../src-base/minecraft/net/minecraft/util/IntHashMap.java
+++ ../src-work/minecraft/net/minecraft/util/IntHashMap.java
@@ -10,7 +10,7 @@
     private int threshold = 12;
     private final float growFactor = 0.75F;
     private transient volatile int versionStamp;
-    private Set keySet = new HashSet();
+    // private Set keySet = new HashSet(); // CraftBukkit - expensive and unused
     private static final String __OBFID = "CL_00001490";
 
     private static int computeHash(int par0)
@@ -61,7 +61,7 @@
 
     public void addKey(int par1, Object par2Obj)
     {
-        this.keySet.add(Integer.valueOf(par1));
+        // this.keySet.add(Integer.valueOf(par1)); // CraftBukkit
         int j = computeHash(par1);
         int k = getSlotIndex(j, this.slots.length);
 
@@ -125,7 +125,7 @@
 
     public Object removeObject(int par1)
     {
-        this.keySet.remove(Integer.valueOf(par1));
+        // this.keySet.remove(Integer.valueOf(par1)); // CraftBukkit
         IntHashMap.Entry entry = this.removeEntry(par1);
         return entry == null ? null : entry.valueEntry;
     }
