--- ../src-base/minecraft/net/minecraft/util/Vec3.java
+++ ../src-work/minecraft/net/minecraft/util/Vec3.java
@@ -7,6 +7,7 @@
     public double xCoord;
     public double yCoord;
     public double zCoord;
+    public Vec3 next; // CraftBukkit
     private static final String __OBFID = "CL_00000612";
 
     public static Vec3 createVectorHelper(double par0, double par2, double par4)
@@ -37,7 +38,7 @@
         this.myVec3LocalPool = par1Vec3Pool;
     }
 
-    protected Vec3 setComponents(double par1, double par3, double par5)
+    public Vec3 setComponents(double par1, double par3, double par5) // CraftBukkit - protected -> public
     {
         this.xCoord = par1;
         this.yCoord = par3;
