--- ../src-base/minecraft/net/minecraft/network/NetworkManager.java
+++ ../src-work/minecraft/net/minecraft/network/NetworkManager.java
@@ -56,7 +56,8 @@
     // JAVADOC FIELD $$ field_150746_k
     private Channel channel;
     // JAVADOC FIELD $$ field_150743_l
-    private SocketAddress socketAddress;
+    public SocketAddress socketAddress; // Spigot
+    public String spoofedUUID; // Spigot
     // JAVADOC FIELD $$ field_150744_m
     private INetHandler netHandler;
     // JAVADOC FIELD $$ field_150741_n
@@ -207,7 +208,15 @@
         {
             for (int i = 1000; !this.receivedPacketsQueue.isEmpty() && i >= 0; --i)
             {
-                Packet packet = (Packet)this.receivedPacketsQueue.poll();
+                Packet packet = (Packet) this.receivedPacketsQueue.poll();
+
+                // CraftBukkit start
+                /*if (!this.isChannelOpen() || !this.channel.config().isAutoRead())   // Should be isConnected
+                {
+                    continue;
+                }*/
+
+                // CraftBukkit end
                 packet.processPacket(this.netHandler);
             }
 
