--- ../src-base/minecraft/net/minecraft/network/NetworkManager.java
+++ ../src-work/minecraft/net/minecraft/network/NetworkManager.java
@@ -51,7 +51,8 @@
     private final Queue receivedPacketsQueue = Queues.newConcurrentLinkedQueue();
     private final Queue outboundPacketsQueue = Queues.newConcurrentLinkedQueue();
     private Channel channel;
-    private SocketAddress socketAddress;
+    public SocketAddress socketAddress; // Spigot
+    public String spoofedUUID; // Spigot
     private INetHandler netHandler;
     private EnumConnectionState connectionState;
     private IChatComponent terminationReason;
@@ -193,7 +194,15 @@
         {
             for (int i = 1000; !this.receivedPacketsQueue.isEmpty() && i >= 0; --i)
             {
-                Packet packet = (Packet)this.receivedPacketsQueue.poll();
+                Packet packet = (Packet) this.receivedPacketsQueue.poll();
+
+                // CraftBukkit start
+                if (!this.isChannelOpen() || !this.channel.config().isAutoRead())   // Should be isConnected
+                {
+                    continue;
+                }
+
+                // CraftBukkit end
                 packet.processPacket(this.netHandler);
             }
 
