--- ../src-base/minecraft/net/minecraft/network/PacketBuffer.java
+++ ../src-work/minecraft/net/minecraft/network/PacketBuffer.java
@@ -97,7 +97,7 @@
 
     public void func_150788_a(ItemStack p_150788_1_) throws IOException
     {
-        if (p_150788_1_ == null)
+        if (p_150788_1_ == null || p_150788_1_.getItem() == null)   // CraftBukkit - NPE fix itemstack.getItem()
         {
             this.writeShort(-1);
         }
