--- ../src-base/minecraft/net/minecraft/network/PingResponseHandler.java
+++ ../src-work/minecraft/net/minecraft/network/PingResponseHandler.java
@@ -30,8 +30,13 @@
 
         try
         {
-            if (bytebuf.readUnsignedByte() == 254)
+            try
             {
+                if (bytebuf.readUnsignedByte() != 254)
+                {
+                    return;
+                }
+
                 InetSocketAddress inetsocketaddress = (InetSocketAddress)p_channelRead_1_.channel().remoteAddress();
                 MinecraftServer minecraftserver = this.field_151257_b.func_151267_d();
                 int i = bytebuf.readableBytes();
@@ -76,13 +81,12 @@
 
                 bytebuf.release();
                 flag = false;
-                return;
             }
+            catch (RuntimeException runtimeexception)
+            {
+                ;
+            }
         }
-        catch (RuntimeException runtimeexception)
-        {
-            return;
-        }
         finally
         {
             if (flag)
