--- ../src-base/minecraft/net/minecraft/item/ItemLilyPad.java
+++ ../src-work/minecraft/net/minecraft/item/ItemLilyPad.java
@@ -34,6 +34,7 @@
                 int i = movingobjectposition.blockX;
                 int j = movingobjectposition.blockY;
                 int k = movingobjectposition.blockZ;
+                final int clickedX = i, clickedY = j, clickedZ = k; // CraftBukkit
 
                 if (!par2World.canMineBlock(par3EntityPlayer, i, j, k))
                 {
@@ -47,8 +48,15 @@
 
                 if (par2World.func_147439_a(i, j, k).func_149688_o() == Material.field_151586_h && par2World.getBlockMetadata(i, j, k) == 0 && par2World.func_147437_c(i, j + 1, k))
                 {
-                    par2World.func_147449_b(i, j + 1, k, Blocks.waterlily);
+                    // CraftBukkit start
+                    // par2World.func_147449_b(i, j + 1, k, Blocks.waterlily);
+                    if (!processBlockPlace(par2World, par3EntityPlayer, null, i, j + 1, k, Blocks.waterlily, 0, clickedX, clickedY, clickedZ))
+                    {
+                        return par1ItemStack;
+                    }
 
+                    // CraftBukkit end
+
                     if (!par3EntityPlayer.capabilities.isCreativeMode)
                     {
                         --par1ItemStack.stackSize;
