--- ../src-base/minecraft/net/minecraft/item/ItemDoor.java
+++ ../src-work/minecraft/net/minecraft/item/ItemDoor.java
@@ -8,6 +8,11 @@
 import net.minecraft.util.MathHelper;
 import net.minecraft.world.World;
 
+// CraftBukkit start
+import net.minecraft.entity.player.EntityPlayerMP;
+import net.minecraft.network.play.server.S23PacketBlockChange;
+// CraftBukkit end
+
 public class ItemDoor extends Item
 {
     private Material doorMaterial;
@@ -23,6 +28,8 @@
     // JAVADOC METHOD $$ func_77648_a
     public boolean onItemUse(ItemStack par1ItemStack, EntityPlayer par2EntityPlayer, World par3World, int par4, int par5, int par6, int par7, float par8, float par9, float par10)
     {
+        final int clickedX = par4, clickedY = par5, clickedZ = par6; // CraftBukkit
+
         if (par7 != 1)
         {
             return false;
@@ -50,7 +57,14 @@
                 else
                 {
                     int i1 = MathHelper.floor_double((double)((par2EntityPlayer.rotationYaw + 180.0F) * 4.0F / 360.0F) - 0.5D) & 3;
-                    func_150924_a(par3World, par4, par5, par6, i1, block);
+
+                    // CraftBukkit start
+                    if (!place(par3World, par4, par5, par6, i1, block, par2EntityPlayer, clickedX, clickedY, clickedZ))
+                    {
+                        return false;
+                    }
+
+                    // CraftBukkit end
                     --par1ItemStack.stackSize;
                     return true;
                 }
@@ -64,33 +78,40 @@
 
     public static void func_150924_a(World p_150924_0_, int p_150924_1_, int p_150924_2_, int p_150924_3_, int p_150924_4_, Block p_150924_5_)
     {
+        // CraftBukkit start
+        place(p_150924_0_, p_150924_1_, p_150924_2_, p_150924_3_, p_150924_4_, p_150924_5_, null, p_150924_1_, p_150924_2_, p_150924_3_);
+    }
+
+    public static boolean place(World world, int i, int j, int k, int l, Block block, EntityPlayer entityplayer, int clickedX, int clickedY, int clickedZ)
+    {
+        // CraftBukkit end
         byte b0 = 0;
         byte b1 = 0;
 
-        if (p_150924_4_ == 0)
+        if (l == 0)
         {
             b1 = 1;
         }
 
-        if (p_150924_4_ == 1)
+        if (l == 1)
         {
             b0 = -1;
         }
 
-        if (p_150924_4_ == 2)
+        if (l == 2)
         {
             b1 = -1;
         }
 
-        if (p_150924_4_ == 3)
+        if (l == 3)
         {
             b0 = 1;
         }
 
-        int i1 = (p_150924_0_.func_147439_a(p_150924_1_ - b0, p_150924_2_, p_150924_3_ - b1).func_149721_r() ? 1 : 0) + (p_150924_0_.func_147439_a(p_150924_1_ - b0, p_150924_2_ + 1, p_150924_3_ - b1).func_149721_r() ? 1 : 0);
-        int j1 = (p_150924_0_.func_147439_a(p_150924_1_ + b0, p_150924_2_, p_150924_3_ + b1).func_149721_r() ? 1 : 0) + (p_150924_0_.func_147439_a(p_150924_1_ + b0, p_150924_2_ + 1, p_150924_3_ + b1).func_149721_r() ? 1 : 0);
-        boolean flag = p_150924_0_.func_147439_a(p_150924_1_ - b0, p_150924_2_, p_150924_3_ - b1) == p_150924_5_ || p_150924_0_.func_147439_a(p_150924_1_ - b0, p_150924_2_ + 1, p_150924_3_ - b1) == p_150924_5_;
-        boolean flag1 = p_150924_0_.func_147439_a(p_150924_1_ + b0, p_150924_2_, p_150924_3_ + b1) == p_150924_5_ || p_150924_0_.func_147439_a(p_150924_1_ + b0, p_150924_2_ + 1, p_150924_3_ + b1) == p_150924_5_;
+        int i1 = (world.func_147439_a(i - b0, j, k - b1).func_149721_r() ? 1 : 0) + (world.func_147439_a(i - b0, j + 1, k - b1).func_149721_r() ? 1 : 0);
+        int j1 = (world.func_147439_a(i + b0, j, k + b1).func_149721_r() ? 1 : 0) + (world.func_147439_a(i + b0, j + 1, k + b1).func_149721_r() ? 1 : 0);
+        boolean flag = world.func_147439_a(i - b0, j, k - b1) == block || world.func_147439_a(i - b0, j + 1, k - b1) == block;
+        boolean flag1 = world.func_147439_a(i + b0, j, k + b1) == block || world.func_147439_a(i + b0, j + 1, k + b1) == block;
         boolean flag2 = false;
 
         if (flag && !flag1)
@@ -102,9 +123,30 @@
             flag2 = true;
         }
 
-        p_150924_0_.func_147465_d(p_150924_1_, p_150924_2_, p_150924_3_, p_150924_5_, p_150924_4_, 2);
-        p_150924_0_.func_147465_d(p_150924_1_, p_150924_2_ + 1, p_150924_3_, p_150924_5_, 8 | (flag2 ? 1 : 0), 2);
-        p_150924_0_.func_147459_d(p_150924_1_, p_150924_2_, p_150924_3_, p_150924_5_);
-        p_150924_0_.func_147459_d(p_150924_1_, p_150924_2_ + 1, p_150924_3_, p_150924_5_);
+        // CraftBukkit start
+        if (entityplayer != null)
+        {
+            if (!ItemBlock.processBlockPlace(world, entityplayer, null, i, j, k, block, l, clickedX, clickedY, clickedZ))
+            {
+                ((EntityPlayerMP) entityplayer).playerNetServerHandler.func_147359_a(new S23PacketBlockChange(i, j + 1, k, world));
+                return false;
+            }
+
+            if (world.func_147439_a(i, j, k) != block)
+            {
+                ((EntityPlayerMP) entityplayer).playerNetServerHandler.func_147359_a(new S23PacketBlockChange(i, j + 1, k, world));
+                return true;
+            }
+        }
+        else
+        {
+            world.func_147465_d(i, j, k, block, l, 2);
+        }
+
+        // CraftBukkit end
+        world.func_147465_d(i, j + 1, k, block, 8 | (flag2 ? 1 : 0), 2);
+        world.func_147459_d(i, j, k, block);
+        world.func_147459_d(i, j + 1, k, block);
+        return true; // CraftBukkit
     }
 }
