--- ../src-base/minecraft/net/minecraft/item/ItemFlintAndSteel.java
+++ ../src-work/minecraft/net/minecraft/item/ItemFlintAndSteel.java
@@ -6,6 +6,11 @@
 import net.minecraft.init.Blocks;
 import net.minecraft.world.World;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.block.CraftBlockState;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+// CraftBukkit end
+
 public class ItemFlintAndSteel extends Item
 {
     private static final String __OBFID = "CL_00000035";
@@ -20,6 +25,8 @@
     // JAVADOC METHOD $$ func_77648_a
     public boolean onItemUse(ItemStack par1ItemStack, EntityPlayer par2EntityPlayer, World par3World, int par4, int par5, int par6, int par7, float par8, float par9, float par10)
     {
+        int clickedX = par4, clickedY = par5, clickedZ = par6; // CraftBukkit
+
         if (par7 == 0)
         {
             --par5;
@@ -58,8 +65,27 @@
         {
             if (par3World.func_147437_c(par4, par5, par6))
             {
+                // CraftBukkit start - Store the clicked block
+                if (CraftEventFactory.callBlockIgniteEvent(par3World, par4, par5, par6, org.bukkit.event.block.BlockIgniteEvent.IgniteCause.FLINT_AND_STEEL, par2EntityPlayer).isCancelled())
+                {
+                    par1ItemStack.damageItem(1, par2EntityPlayer);
+                    return false;
+                }
+
+                CraftBlockState blockState = CraftBlockState.getBlockState(par3World, par4, par5, par6);
+                // CraftBukkit end
                 par3World.playSoundEffect((double)par4 + 0.5D, (double)par5 + 0.5D, (double)par6 + 0.5D, "fire.ignite", 1.0F, itemRand.nextFloat() * 0.4F + 0.8F);
                 par3World.func_147449_b(par4, par5, par6, Blocks.fire);
+                // CraftBukkit start
+                org.bukkit.event.block.BlockPlaceEvent placeEvent = CraftEventFactory.callBlockPlaceEvent(par3World, par2EntityPlayer, blockState, clickedX, clickedY, clickedZ);
+
+                if (placeEvent.isCancelled() || !placeEvent.canBuild())
+                {
+                    placeEvent.getBlockPlaced().setTypeIdAndData(0, (byte) 0, false);
+                    return false;
+                }
+
+                // CraftBukkit end
             }
 
             par1ItemStack.damageItem(1, par2EntityPlayer);
