--- ../src-base/minecraft/cpw/mods/fml/common/FMLCommonHandler.java
+++ ../src-work/minecraft/cpw/mods/fml/common/FMLCommonHandler.java
@@ -495,7 +495,12 @@
 
     public String getModName()
     {
-        List<String> modNames = Lists.newArrayListWithExpectedSize(3);
+        // Cauldron start    
+        List<String> modNames = Lists.newArrayListWithExpectedSize(6);
+        modNames.add("cauldron");
+        modNames.add("craftbukkit");
+        modNames.add("mcpc");
+        // Cauldron end
         modNames.add("fml");
         if (!noForge)
         {
