--- ../src-base/minecraft/net/minecraft/network/play/server/S26PacketMapChunkBulk.java
+++ ../src-work/minecraft/net/minecraft/network/play/server/S26PacketMapChunkBulk.java
@@ -13,6 +13,7 @@
 import net.minecraft.network.PacketBuffer;
 import net.minecraft.network.play.INetHandlerPlayClient;
 import net.minecraft.world.chunk.Chunk;
+import net.minecraft.world.World; // Spigot
 
 public class S26PacketMapChunkBulk extends Packet
 {
@@ -24,10 +25,19 @@
     private byte[][] field_149260_f;
     private int field_149261_g;
     private boolean field_149267_h;
-    private static byte[] field_149268_i = new byte[0];
-    private static final String __OBFID = "CL_00001306";
-    private int maxLen = 0;
-    private Semaphore deflateGate;
+    private byte[] field_149268_i = new byte[0]; // CraftBukkit - remove static
+    // CraftBukkit start
+    static final ThreadLocal<Deflater> localDeflater = new ThreadLocal<Deflater>()
+    {
+        @Override
+        protected Deflater initialValue()
+        {
+            // Don't use higher compression level, slows things down too much
+            return new Deflater(4); // Spigot 6 -> 4
+        }
+    };
+    // CraftBukkit end
+    private World world; // Spigot
 
     public S26PacketMapChunkBulk() {}
 
@@ -46,6 +56,7 @@
         {
             Chunk chunk = (Chunk)p_i45197_1_.get(k);
             S21PacketChunkData.Extracted extracted = S21PacketChunkData.func_149269_a(chunk, true, 65535);
+            world = chunk.worldObj; // Spigot
             j += extracted.field_150282_a.length;
             this.field_149266_a[k] = chunk.xPosition;
             this.field_149264_b[k] = chunk.zPosition;
@@ -53,34 +64,57 @@
             this.field_149262_d[k] = extracted.field_150281_c;
             this.field_149260_f[k] = extracted.field_150282_a;
         }
-        this.deflateGate = new Semaphore(1);
-        maxLen = j;
+        /* CraftBukkit start - Moved to compress()
+        Deflater deflater = new Deflater(-1);
+
+        try {
+            deflater.setInput(buildBuffer, 0, j);
+            deflater.finish();
+            this.buffer = new byte[j];
+            this.size = deflater.deflate(this.buffer);
+        } finally {
+            deflater.end();
+        }
+        */
     }
 
-    private void deflate()
+    // Add compression method
+    public void compress()
     {
-        byte[] data = new byte[maxLen];
-        int offset = 0;
-        for (int x = 0; x < field_149260_f.length; x++)
+        if (this.field_149263_e != null)
         {
-            System.arraycopy(field_149260_f[x], 0, data, offset, field_149260_f[x].length);
-            offset += field_149260_f[x].length;
+            return;
         }
-        Deflater deflater = new Deflater(-1);
 
-        try
+        // Spigot start
+        int finalBufferSize = 0;
+
+        // Obfuscate all sections
+        for (int i = 0; i < field_149266_a.length; i++)
         {
-            deflater.setInput(data, 0, data.length);
-            deflater.finish();
-            byte[] deflated = new byte[data.length];
-            this.field_149261_g = deflater.deflate(deflated);
-            this.field_149263_e = deflated;
+            world.spigotConfig.antiXrayInstance.obfuscate(field_149266_a[i], field_149264_b[i], field_149265_c[i], field_149260_f[i], world);
+            finalBufferSize += field_149260_f[i].length;
         }
-        finally
+
+        // Now it's time to efficiently copy the chunk to the build buffer
+        field_149268_i = new byte[finalBufferSize];
+        int bufferLocation = 0;
+
+        for (int i = 0; i < field_149266_a.length; i++)
         {
-            deflater.end();
+            System.arraycopy(field_149260_f[i], 0, field_149268_i, bufferLocation, field_149260_f[i].length);
+            bufferLocation += field_149260_f[i].length;
         }
+
+        // Spigot end
+        Deflater deflater = localDeflater.get();
+        deflater.reset();
+        deflater.setInput(this.field_149268_i);
+        deflater.finish();
+        this.field_149263_e = new byte[this.field_149268_i.length + 100];
+        this.field_149261_g = deflater.deflate(this.field_149263_e);
     }
+    // CraftBukkit end
 
     public static int func_149258_c()
     {
@@ -155,16 +189,7 @@
 
     public void func_148840_b(PacketBuffer p_148840_1_) throws IOException
     {
-        if (this.field_149263_e == null)
-        {
-            deflateGate.acquireUninterruptibly();
-            if (this.field_149263_e == null)
-            {
-                deflate();
-            }
-            deflateGate.release();
-        }
-
+        compress(); // CraftBukkit
         p_148840_1_.writeShort(this.field_149266_a.length);
         p_148840_1_.writeInt(this.field_149261_g);
         p_148840_1_.writeBoolean(this.field_149267_h);
