--- ../src-base/minecraft/net/minecraft/client/gui/GuiMultiplayer.java
+++ ../src-work/minecraft/net/minecraft/client/gui/GuiMultiplayer.java
@@ -2,7 +2,6 @@
 
 import com.google.common.base.Splitter;
 import com.google.common.collect.Lists;
-import cpw.mods.fml.client.FMLClientHandler;
 import cpw.mods.fml.relauncher.Side;
 import cpw.mods.fml.relauncher.SideOnly;
 import java.util.List;
@@ -41,7 +40,6 @@
     public GuiMultiplayer(GuiScreen par1GuiScreen)
     {
         this.field_146798_g = par1GuiScreen;
-        FMLClientHandler.instance().setupServerList();
     }
 
     // JAVADOC METHOD $$ func_73866_w_
