--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenSwamp.java
+++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenSwamp.java
@@ -8,7 +8,9 @@
 import net.minecraft.world.World;
 import net.minecraftforge.common.util.ForgeDirection;
 
-public class WorldGenSwamp extends WorldGenAbstractTree
+import org.bukkit.craftbukkit.CraftBlockChangeDelegate; // CraftBukkit
+
+public class WorldGenSwamp extends WorldGenAbstractTree implements BlockSapling.TreeGenerator   // CraftBukkit - add interface
 {
     private static final String __OBFID = "CL_00000436";
 
@@ -19,9 +21,16 @@
 
     public boolean generate(World par1World, Random par2Random, int par3, int par4, int par5)
     {
+        // CraftBukkit start - Moved to generate
+        return this.generate(new CraftBlockChangeDelegate((org.bukkit.BlockChangeDelegate) par1World), par2Random, par3, par4, par5);
+    }
+
+    public boolean generate(CraftBlockChangeDelegate par1World, Random par2Random, int par3, int par4, int par5)
+    {
+        // CraftBukkit end
         int l;
 
-        for (l = par2Random.nextInt(4) + 5; par1World.func_147439_a(par3, par4 - 1, par5).func_149688_o() == Material.field_151586_h; --par4)
+        for (l = par2Random.nextInt(4) + 5; par1World.getType(par3, par4 - 1, par5).func_149688_o() == Material.field_151586_h; --par4)
         {
             ;
         }
@@ -53,9 +62,9 @@
                     {
                         if (i1 >= 0 && i1 < 256)
                         {
-                            Block block = par1World.func_147439_a(j1, i1, k1);
+                            Block block = par1World.getType(j1, i1, k1);
 
-                            if (!(block.isAir(par1World, j1, i1, k1) || block.isLeaves(par1World, j1, i1, k1)))
+                            if (block.isAir(par1World, j1, i1, k1) || block.isLeaves(par1World, j1, i1, k1))
                             {
                                 if (block != Blocks.water && block != Blocks.flowing_water)
                                 {
@@ -81,7 +90,7 @@
             }
             else
             {
-                Block block1 = par1World.func_147439_a(par3, par4 - 1, par5);
+                Block block1 = par1World.getType(par3, par4 - 1, par5);
 
                 boolean isSoil = block1.canSustainPlant(par1World, par3, par4 - 1, par5, ForgeDirection.UP, (BlockSapling)Blocks.sapling);
                 if (isSoil && par4 < 256 - l - 1)
@@ -104,9 +113,9 @@
                             {
                                 int j2 = i2 - par5;
 
-                                if ((Math.abs(l1) != k1 || Math.abs(j2) != k1 || par2Random.nextInt(2) != 0 && j1 != 0) && par1World.func_147439_a(l2, k2, i2).canBeReplacedByLeaves(par1World, l2, k2, i2))
+                                if ((Math.abs(l1) != k1 || Math.abs(j2) != k1 || par2Random.nextInt(2) != 0 && j1 != 0) && !par1World.getType(l2, k2, i2).canBeReplacedByLeaves(par1World, l2, k2, i2))
                                 {
-                                    this.func_150515_a(par1World, l2, k2, i2, Blocks.leaves);
+                                    this.setType(par1World, l2, k2, i2, Blocks.leaves);
                                 }
                             }
                         }
@@ -114,11 +123,11 @@
 
                     for (k2 = 0; k2 < l; ++k2)
                     {
-                        Block block2 = par1World.func_147439_a(par3, par4 + k2, par5);
+                        Block block2 = par1World.getType(par3, par4 + k2, par5);
 
                         if (block2.isAir(par1World, par3, par4 + k2, par5) || block2.isLeaves(par1World, par3, par4 + k2, par5) || block2 == Blocks.flowing_water || block2 == Blocks.water)
                         {
-                            this.func_150515_a(par1World, par3, par4 + k2, par5, Blocks.log);
+                            this.setType(par1World, par3, par4 + k2, par5, Blocks.log);
                         }
                     }
 
@@ -131,24 +140,24 @@
                         {
                             for (l1 = par5 - k1; l1 <= par5 + k1; ++l1)
                             {
-                                if (par1World.func_147439_a(l2, k2, l1).isLeaves(par1World, l2, k2, l1))
+                                if (par1World.getType(l2, k2, l1).isLeaves(par1World, l2, k2, l1))
                                 {
-                                    if (par2Random.nextInt(4) == 0 && par1World.func_147439_a(l2 - 1, k2, l1).isAir(par1World, l2 - 1, k2, l1))
+                                    if (par2Random.nextInt(4) == 0 && par1World.getType(l2 - 1, k2, l1).isAir(par1World, l2 - 1, k2, l1))
                                     {
                                         this.generateVines(par1World, l2 - 1, k2, l1, 8);
                                     }
 
-                                    if (par2Random.nextInt(4) == 0 && par1World.func_147439_a(l2 + 1, k2, l1).isAir(par1World, l2 + 1, k2, l1))
+                                    if (par2Random.nextInt(4) == 0 && par1World.getType(l2 + 1, k2, l1).isAir(par1World, l2 + 1, k2, l1))
                                     {
                                         this.generateVines(par1World, l2 + 1, k2, l1, 2);
                                     }
 
-                                    if (par2Random.nextInt(4) == 0 && par1World.func_147439_a(l2, k2, l1 - 1).isAir(par1World, l2, k2, l1 - 1))
+                                    if (par2Random.nextInt(4) == 0 && par1World.getType(l2, k2, l1 - 1).isAir(par1World, l2, k2, l1 - 1))
                                     {
                                         this.generateVines(par1World, l2, k2, l1 - 1, 1);
                                     }
 
-                                    if (par2Random.nextInt(4) == 0 && par1World.func_147439_a(l2, k2, l1 + 1).isAir(par1World, l2, k2, l1 + 1))
+                                    if (par2Random.nextInt(4) == 0 && par1World.getType(l2, k2, l1 + 1).isAir(par1World, l2, k2, l1 + 1))
                                     {
                                         this.generateVines(par1World, l2, k2, l1 + 1, 4);
                                     }
@@ -172,21 +181,22 @@
     }
 
     // JAVADOC METHOD $$ func_76536_b
-    private void generateVines(World par1World, int par2, int par3, int par4, int par5)
+    // CraftBukkit - change signature
+    private void generateVines(CraftBlockChangeDelegate par1World, int par2, int par3, int par4, int par5)
     {
-        this.func_150516_a(par1World, par2, par3, par4, Blocks.vine, par5);
+        this.setTypeAndData(par1World, par2, par3, par4, Blocks.vine, par5);
         int i1 = 4;
 
         while (true)
         {
             --par3;
 
-            if (!(par1World.func_147439_a(par2, par3, par4).isAir(par1World, par2, par3, par4)) || i1 <= 0)
+            if (par1World.getType(par2, par3, par4).isAir(par1World, par2, par3, par4) || i1 <= 0)
             {
                 return;
             }
 
-            this.func_150516_a(par1World, par2, par3, par4, Blocks.vine, par5);
+            this.setTypeAndData(par1World, par2, par3, par4, Blocks.vine, par5);
             --i1;
         }
     }
