--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenSavannaTree.java
+++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenSavannaTree.java
@@ -9,7 +9,9 @@
 import net.minecraft.world.World;
 import net.minecraftforge.common.util.ForgeDirection;
 
-public class WorldGenSavannaTree extends WorldGenAbstractTree
+import org.bukkit.craftbukkit.CraftBlockChangeDelegate; // CraftBukkit
+
+public class WorldGenSavannaTree extends WorldGenAbstractTree implements BlockSapling.TreeGenerator   // CraftBukkit - add interface
 {
     private static final String __OBFID = "CL_00000432";
 
@@ -20,6 +22,13 @@
 
     public boolean generate(World par1World, Random par2Random, int par3, int par4, int par5)
     {
+        // CraftBukkit start - Moved to generate
+        return this.generate(new CraftBlockChangeDelegate((org.bukkit.BlockChangeDelegate) par1World), par2Random, par3, par4, par5);
+    }
+
+    public boolean generate(CraftBlockChangeDelegate par1World, Random par2Random, int par3, int par4, int par5)
+    {
+        // CraftBukkit end
         int l = par2Random.nextInt(3) + par2Random.nextInt(3) + 5;
         boolean flag = true;
 
@@ -48,7 +57,7 @@
                     {
                         if (i1 >= 0 && i1 < 256)
                         {
-                            Block block = par1World.func_147439_a(j1, i1, k1);
+                            Block block = par1World.getType(j1, i1, k1);
 
                             if (!this.isReplaceable(par1World, j1, i1, k1))
                             {
@@ -69,7 +78,7 @@
             }
             else
             {
-                Block block3 = par1World.func_147439_a(par3, par4 - 1, par5);
+                Block block3 = par1World.getType(par3, par4 - 1, par5);
 
                 boolean isSoil = block3.canSustainPlant(par1World, par3, par4 - 1, par5, ForgeDirection.UP, (BlockSapling)Blocks.sapling);
                 if (isSoil && par4 < 256 - l - 1)
@@ -95,11 +104,11 @@
                             --k1;
                         }
 
-                        Block block1 = par1World.func_147439_a(k3, k2, l1);
+                        Block block1 = par1World.getType(k3, k2, l1);
 
                         if (block1.isAir(par1World, k3, k2, l1) || block1.isLeaves(par1World, k3, k2, l1))
                         {
-                            this.func_150516_a(par1World, k3, k2, l1, Blocks.log2, 0);
+                            this.setTypeAndData(par1World, k3, k2, l1, Blocks.log2, 0);
                             i2 = k2;
                         }
                     }
@@ -147,11 +156,11 @@
                                 i3 = par4 + l2;
                                 k3 += Direction.offsetX[j2];
                                 l1 += Direction.offsetZ[j2];
-                                Block block2 = par1World.func_147439_a(k3, i3, l1);
+                                Block block2 = par1World.getType(k3, i3, l1);
 
                                 if (block2.isAir(par1World, k3, i3, l1) || block2.isLeaves(par1World, k3, i3, l1))
                                 {
-                                    this.func_150516_a(par1World, k3, i3, l1, Blocks.log2, 0);
+                                    this.setTypeAndData(par1World, k3, i3, l1, Blocks.log2, 0);
                                     i2 = i3;
                                 }
                             }
@@ -196,13 +205,14 @@
         }
     }
 
-    private void func_150525_a(World p_150525_1_, int p_150525_2_, int p_150525_3_, int p_150525_4_)
+    // CraftBukkit - Changed signature
+    private void func_150525_a(CraftBlockChangeDelegate p_150525_1_, int p_150525_2_, int p_150525_3_, int p_150525_4_)
     {
-        Block block = p_150525_1_.func_147439_a(p_150525_2_, p_150525_3_, p_150525_4_);
+        Block block = p_150525_1_.getType(p_150525_2_, p_150525_3_, p_150525_4_);
 
         if (block.isAir(p_150525_1_, p_150525_2_, p_150525_3_, p_150525_4_) || block.isLeaves(p_150525_1_, p_150525_2_, p_150525_3_, p_150525_4_))
         {
-            this.func_150516_a(p_150525_1_, p_150525_2_, p_150525_3_, p_150525_4_, Blocks.leaves2, 0);
+            this.setTypeAndData(p_150525_1_, p_150525_2_, p_150525_3_, p_150525_4_, Blocks.leaves2, 0);
         }
     }
 }
