--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenerator.java
+++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenerator.java
@@ -4,6 +4,8 @@
 import net.minecraft.block.Block;
 import net.minecraft.world.World;
 
+import org.bukkit.craftbukkit.CraftBlockChangeDelegate; // CraftBukkit
+
 public abstract class WorldGenerator
 {
     // JAVADOC FIELD $$ field_76488_a
@@ -27,11 +29,18 @@
 
     protected void func_150515_a(World p_150515_1_, int p_150515_2_, int p_150515_3_, int p_150515_4_, Block p_150515_5_)
     {
-        this.func_150516_a(p_150515_1_, p_150515_2_, p_150515_3_, p_150515_4_, p_150515_5_, 0);
+        this.setTypeAndData(p_150515_1_, p_150515_2_, p_150515_3_, p_150515_4_, p_150515_5_, 0);
     }
 
-    protected void func_150516_a(World p_150516_1_, int p_150516_2_, int p_150516_3_, int p_150516_4_, Block p_150516_5_, int p_150516_6_)
+    // CraftBukkit start - Duplicate method to add support for CraftBlockChangeDelegate
+    protected void setType(CraftBlockChangeDelegate world, int i, int j, int k, Block block)
     {
+        this.setTypeAndData(world, i, j, k, block, 0);
+    }
+    // CraftBukkit end
+
+    protected void setTypeAndData(World p_150516_1_, int p_150516_2_, int p_150516_3_, int p_150516_4_, Block p_150516_5_, int p_150516_6_)
+    {
         if (this.doBlockNotify)
         {
             p_150516_1_.func_147465_d(p_150516_2_, p_150516_3_, p_150516_4_, p_150516_5_, p_150516_6_, 3);
@@ -41,4 +50,18 @@
             p_150516_1_.func_147465_d(p_150516_2_, p_150516_3_, p_150516_4_, p_150516_5_, p_150516_6_, 2);
         }
     }
+
+    // CraftBukkit start - Duplicate method to add support for CraftBlockChangeDelegate
+    protected void setTypeAndData(CraftBlockChangeDelegate world, int i, int j, int k, Block block, int l)
+    {
+        if (this.doBlockNotify)
+        {
+            world.setTypeAndData(i, j, k, block, l, 3);
+        }
+        else
+        {
+            world.setTypeAndData(i, j, k, block, l, 2);
+        }
+    }
+    // CraftBukkit end
 }
