--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenTaiga1.java
+++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenTaiga1.java
@@ -8,7 +8,9 @@
 import net.minecraft.world.World;
 import net.minecraftforge.common.util.ForgeDirection;
 
-public class WorldGenTaiga1 extends WorldGenAbstractTree
+import org.bukkit.craftbukkit.CraftBlockChangeDelegate; // CraftBukkit
+
+public class WorldGenTaiga1 extends WorldGenAbstractTree implements BlockSapling.TreeGenerator   // CraftBukkit - add interface
 {
     private static final String __OBFID = "CL_00000427";
 
@@ -19,6 +21,13 @@
 
     public boolean generate(World par1World, Random par2Random, int par3, int par4, int par5)
     {
+        // CraftBukkit start - Moved to generate
+        return this.generate(new CraftBlockChangeDelegate((org.bukkit.BlockChangeDelegate) par1World), par2Random, par3, par4, par5);
+    }
+
+    public boolean generate(CraftBlockChangeDelegate par1World, Random par2Random, int par3, int par4, int par5)
+    {
+        // CraftBukkit end
         int l = par2Random.nextInt(5) + 7;
         int i1 = l - par2Random.nextInt(2) - 3;
         int j1 = l - i1;
@@ -50,7 +59,7 @@
                     {
                         if (l1 >= 0 && l1 < 256)
                         {
-                            Block block = par1World.func_147439_a(i2, l1, j2);
+                            Block block = par1World.getType(i2, l1, j2);
 
                             if (!this.isReplaceable(par1World, i2, l1, j2))
                             {
@@ -71,7 +80,7 @@
             }
             else
             {
-                Block block1 = par1World.func_147439_a(par3, par4 - 1, par5);
+                Block block1 = par1World.getType(par3, par4 - 1, par5);
 
                 boolean isSoil = block1.canSustainPlant(par1World, par3, par4 - 1, par5, ForgeDirection.UP, (BlockSapling)Blocks.sapling);
                 if (isSoil && par4 < 256 - l - 1)
@@ -89,9 +98,9 @@
                             {
                                 int l2 = k2 - par5;
 
-                                if ((Math.abs(j3) != i3 || Math.abs(l2) != i3 || i3 <= 0) && par1World.func_147439_a(j2, i2, k2).canBeReplacedByLeaves(par1World, j2, i2, k2))
+                                if ((Math.abs(j3) != i3 || Math.abs(l2) != i3 || i3 <= 0) && !par1World.getType(j2, i2, k2).canBeReplacedByLeaves(par1World, j2, i2, k2))
                                 {
-                                    this.func_150516_a(par1World, j2, i2, k2, Blocks.leaves, 1);
+                                    this.setTypeAndData(par1World, j2, i2, k2, Blocks.leaves, 1);
                                 }
                             }
                         }
@@ -108,11 +117,11 @@
 
                     for (i2 = 0; i2 < l - 1; ++i2)
                     {
-                        Block block2 = par1World.func_147439_a(par3, par4 + i2, par5);
+                        Block block2 = par1World.getType(par3, par4 + i2, par5);
 
                         if (block2.isAir(par1World, par3, par4 + i2, par5) || block2.isLeaves(par1World, par3, par4 + i2, par5))
                         {
-                            this.func_150516_a(par1World, par3, par4 + i2, par5, Blocks.log, 1);
+                            this.setTypeAndData(par1World, par3, par4 + i2, par5, Blocks.log, 1);
                         }
                     }
 
