--- ../src-base/minecraft/net/minecraftforge/common/ForgeVersion.java
+++ ../src-work/minecraft/net/minecraftforge/common/ForgeVersion.java
@@ -102,7 +102,6 @@
                     {
                         ArtifactVersion recommended = new DefaultArtifactVersion(rec);
                         int diff = recommended.compareTo(current);
-
                         if (diff == 0)
                             status = UP_TO_DATE;
                         else if (diff < 0)
@@ -145,4 +144,3 @@
         }.start();
     }
 }
-
